---
title: "Analysis"
author: "Daniel Hanhart"
date: "3/16/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scLinear)
library(tidyverse)
library(ggplot2)
library(Seurat)
library(svglite)
library(Metrics)
library(pals)
library(ggpubr)
library(DropletUtils)

set.seed(42)
```

## Mouse lymphnode spleen

### Create Seurat object from raw data

```{r}
# create seurat object from pbmc5k data
## Some of the protein names in pbmc5k differ from the ones used in the NeurIPS data (CD8a -> CD8, PD-1 -> CD279). Create name mapping."
# how to compare CD21_CD35 vs CD21 and CD35?
 map_names <- data.frame(
 	mouse = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226_10E5", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R_B220", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90.2", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype_G0114F7", "Isotype_HTK888", "Isotype_MOPC.173", "Isotype_MOPC.21", "Isotype_MPC.11", "Isotype_RTK2071", "Isotype_RTK2758", "Isotype_RTK4174",
 	          "Isotype_RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "Ly108", "Ly49D", "Ly49H", "NK.1.1", "Siglec_H", "TCR.Bchain", "TCR.GD_UC7", "TCR.RD_GL3", "TCR.Va2", "TCR.Va8.3_KT50", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA"),
 	
neurips = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90.2", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype_G0114F7", "Isotype_HTK888", "Isotype_MOPC.173", "Isotype_MOPC.21", "Isotype_MPC.11", "Isotype_RTK2071", "Isotype_RTK2758", "Isotype_RTK4174",
            "Isotype_RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "CD352", "Ly49D", "Ly49H", "CD161", "Siglec_H", "TCR.Bchain", "TCR.GD", "TCR.RD", "TCR.Va2", "TCR.Va8.3", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA")
 )

mls.data <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_Multiplex_count_raw_feature_bc_matrix.h5")

## remove HuMsRt. for human, mouse and rat binding antibodies
rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^HuMsRt\\.", replacement = "",
                                                           x = rownames(x = mls.data[["Antibody Capture"]]))
## remove HuMs. for human and mouse binding antibodies
rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^HuMs\\.", replacement = "",
                                                           x = rownames(x = mls.data[["Antibody Capture"]]))
## remove MsRt. for human and mouse binding antibodies
rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^MsRt\\.", replacement = "",
                                                           x = rownames(x = mls.data[["Antibody Capture"]]))
## remove Ms. for mouse binding antibodies
rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^Ms\\.", replacement = "",
                                                           x = rownames(x = mls.data[["Antibody Capture"]]))

rownames(x = mls.data[["Antibody Capture"]]) <- map_names$neurips[ match(rownames(x = mls.data[["Antibody Capture"]]), map_names$mouse)]

mls <- Seurat::CreateSeuratObject(counts = mls.data[["Gene Expression"]], min.cells = 0, min.features = 0)
mls[["ADT"]] <- Seurat::CreateAssayObject(mls.data[["Antibody Capture"]][, colnames(x = mls)])
Seurat::DefaultAssay(mls) <- "RNA"

mls$mito_percent <- Seurat::PercentageFeatureSet(object = mls, pattern = "^mt-")

saveRDS(mls, "./../local/mls_raw.rds")
rm(mls.data, mls, map_names)
gc()
```

### Preprocess MLS data

```{r}
mls <- readRDS("./../local/mls_raw.rds")

mls <- scLinear::empty_drops(object = mls, lower = 100, FDR = 0.01, samples = NULL )


mls <- remove_doublets(mls, samples = NULL, print_plots = TRUE)

mls <- mad_filtering(mls$object, samples = NULL, print_plots = TRUE)
mls <- mls$object

mls <- cluster_data(mls, resolution = 0.8)

DimPlot(mls, group.by = "seurat_clusters", pt.size = 0.05)

a <- mls %>% anno_celltypes(anno_level = 2, selfClusters = Seurat::Idents(.) ,species = "Mm")

VlnPlot(mls, features = c("nFeature_RNA", "nCount_RNA", "mito_percent"), ncol = 3, pt.size = 0)


## Perform basc preprocessing steps including: lognormalization, removal of doublets, removal of low quality cells, clustering, cell-type annotation
pbmc5k <- scLinear::prepare_data(pbmc5k,
                                   remove_doublets = TRUE,
                                   low_qc_cell_removal = TRUE,
                                   anno_level = c(2,3),
                                   samples = NULL,
                                   integrate_data = FALSE,
                                   remove_empty_droplets = FALSE,
                                   lower = 100, FDR = 0.01,
                                   annotation_selfCluster = TRUE,
                                   resolution = 0.8,
                                   seed = 42)

saveRDS(pbmc5k, "./../local/mls_prepared.rds")
rm(pbmc5k)
gc()
```
### Take a look into the ADT data
```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared.rds")

adt_counts <- GetAssayData(pbmc5k, slot = "counts", assay = "ADT")

iso_features <- c("IgG1", "IgG2a", "IgG2b")
all_features <- rownames(pbmc5k@assays$ADT)
iso_count <- colSums(x = GetAssayData(object = pbmc5k, assay = "ADT", slot = "counts")[iso_features, , drop = FALSE])
total_counts_ADT <- colSums(x = GetAssayData(object = pbmc5k, assay = "ADT", slot = "counts")[all_features, , drop = FALSE])


qc.stats <- DropletUtils::cleanTagCounts(adt_counts, controls = iso_features) %>% as.data.frame()

high_control_threshold <- attr(qc.stats$high.controls, "thresholds")

DF <- data.frame(total_counts = total_counts_ADT, iso_counts = iso_count, nFeatures = pbmc5k$nFeature_ADT)

p<-ggplot(DF, aes(x=total_counts_ADT, y=nFeatures, color = iso_count)) + 
  geom_point(binaxis='y', stackdir='center') + theme_bw() +
  scale_color_gradient(low = "blue", high = "red", trans = "log10") + 
  guides(colour = guide_colourbar(barheight = unit(8, "cm")))
p


## 
DF2 <- DF %>% pivot_longer(!nFeatures, values_to = "counts", names_to = "names")
p <- ggplot(DF2, aes(x = counts, fill = names)) + geom_histogram() +
  theme_bw() +
  scale_fill_manual(values = kelly()[3:length(kelly())]) +
  facet_wrap(~names , ncol = 1, scales = "free") +
  geom_vline(aes(xintercept = xintercept, color = Lines), data.frame(xintercept = high_control_threshold["higher"], Lines = "upper_iso_count_threshold") , linetype = "dashed")

p

## Normalize ADT data
pbmc5k <- Seurat::NormalizeData(pbmc5k, normalization.method = "CLR", margin = 2, assay = "ADT")

saveRDS(pbmc5k, "./../local/PBMC5K_prepared_2.rds")

rm(pbmc5k, adt_counts, DF, DF2, p, qc.stats, all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT)
gc()
```
### UMAP

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared_2.rds")

p <- DimPlot(pbmc5k, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
  scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300")) + theme_bw()
p
ggsave(filename = "./../local/umap_pbmc5k.pdf", plot = p, width = 9, height = 6, units = "cm")
rm(p, pbmc5k)
gc()
```

### predict PBMC5K with scLinear

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared_2.rds")

## load pretrained predictor model scLinear
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")
assay_predicted <- adt_predict(pipe, pbmc5k@assays$RNA, normalize = TRUE)

eval_res <- evaluate_predictor(pipe, pbmc5k@assays$RNA, pbmc5k@assays$ADT)

pbmc5k[["predicted_adt_sclinear"]] <- assay_predicted

saveRDS(pbmc5k, "./../local/PBMC5K_predicted.rds")
rm(assay_predicted, eval_res,pipe)
gc()
```

### bit table (RNA / protein predicted vs protein real)

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_predicted.rds")

adt_real <- as.matrix(pbmc5k@assays[["ADT"]]@data)
adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
adt_predicted_sclinear <- as.matrix(pbmc5k@assays[["predicted_adt_sclinear"]]@data)
adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_sclinear")

#### add raw RNA
rna_raw <- as.matrix(pbmc5k@assays[["RNA"]]@counts)
rna_raw <- rna_raw %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_raw")
#### post processing of RNA data to make it comparable to ADT data
rna_raw$gene2 <- rna_raw$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_raw$gene2[rna_raw$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_raw$gene2[rna_raw$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_raw$gene2[rna_raw$gene %in% c("ITGAM")] <- "CD11b"
rna_raw$gene2[rna_raw$gene %in% c("FUT4")] <- "CD15"
rna_raw$gene2[rna_raw$gene %in% c("FCGR3A")] <- "CD16"
rna_raw$gene2[rna_raw$gene %in% c("MS4A1")] <- "CD20"
rna_raw$gene2[rna_raw$gene %in% c("IL2RA")] <- "CD25"
rna_raw$gene2[rna_raw$gene %in% c("NCAM1")] <- "CD56"
rna_raw$gene2[rna_raw$gene %in% c("SELL")] <- "CD62L"
rna_raw$gene2[rna_raw$gene %in% c("IL7R")] <- "CD127"
rna_raw$gene2[rna_raw$gene %in% c("TNFR5F9")] <- "CD137"
rna_raw$gene2[rna_raw$gene %in% c("CCR7")] <- "CD197"
rna_raw$gene2[rna_raw$gene %in% c("ICOS")] <- "CD278"
rna_raw$gene2[rna_raw$gene %in% c("NCR1")] <- "CD335"
rna_raw$gene2[rna_raw$gene %in% c("PDCD1")] <- "CD279"
## three genes part of the HLA-DR complex are measures (take mean at the end)
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRA")] <- "HLA-DR"
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRB1")] <- "HLA-DR"
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRB5")] <- "HLA-DR"


### PTPRC -> CD45RA / CD45RO (ignored for nown not clear to which isoform the gene counts should be compared to!)
#-
rna_raw <- rna_raw %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_raw") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_raw <- rna_raw %>% group_by(cell, gene) %>% summarise(rna_raw = mean(rna_raw))

#### add normalized RNA (normalized the same way as used for prediction)
rna_real <- as.matrix(scLinear::gexp_normalize(pbmc5k@assays[["RNA"]]@counts))
rna_real <- rna_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_normalized")
#### post processing of RNA data to make it comparable to ADT data
rna_real$gene2 <- rna_real$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_real$gene2[rna_real$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_real$gene2[rna_real$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_real$gene2[rna_real$gene %in% c("ITGAM")] <- "CD11b"
rna_real$gene2[rna_real$gene %in% c("FUT4")] <- "CD15"
rna_real$gene2[rna_real$gene %in% c("FCGR3A")] <- "CD16"
rna_real$gene2[rna_real$gene %in% c("MS4A1")] <- "CD20"
rna_real$gene2[rna_real$gene %in% c("IL2RA")] <- "CD25"
rna_real$gene2[rna_real$gene %in% c("NCAM1")] <- "CD56"
rna_real$gene2[rna_real$gene %in% c("SELL")] <- "CD62L"
rna_real$gene2[rna_real$gene %in% c("IL7R")] <- "CD127"
rna_real$gene2[rna_real$gene %in% c("TNFR5F9")] <- "CD137"
rna_real$gene2[rna_real$gene %in% c("CCR7")] <- "CD197"
rna_real$gene2[rna_real$gene %in% c("ICOS")] <- "CD278"
rna_real$gene2[rna_real$gene %in% c("NCR1")] <- "CD335"
rna_real$gene2[rna_real$gene %in% c("PDCD1")] <- "CD279"
## three genes part of the HLA-DR complex are measures (take mean at the end)
rna_real$gene2[rna_real$gene %in% c("HLA-DRA")] <- "HLA-DR"
rna_real$gene2[rna_real$gene %in% c("HLA-DRB1")] <- "HLA-DR"
rna_real$gene2[rna_real$gene %in% c("HLA-DRB5")] <- "HLA-DR"

### PTPRC -> CD45RA / CD45RO (ignored for nown not clear to which isoform the gene counts should be compared to!)
#-
rna_real <- rna_real %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_normalized") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_real <- rna_real %>% group_by(cell, gene) %>% summarise(rna_normalized = mean(rna_normalized))

meta <- pbmc5k@meta.data %>% rownames_to_column("cell") %>% dplyr::select("cell","cell_type_2","cell_type_3")

DF <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>%  full_join(rna_real, by = c("gene", "cell")) %>% full_join(rna_raw, by = c("gene", "cell")) %>% full_join(meta, by = c("cell"))

write.table(DF, file = "./../local/PBMC5K/bit_table_pbmc5k_2.csv", sep = ",", col.names = TRUE, row.names = FALSE)
```



```{r, fig.width= 15, fig.height = 10}
DF <- read.table("./../local/PBMC5K/bit_table_pbmc5k_2.csv", header = T, sep=',')
d <- DF %>% na.omit() %>% group_by(gene)  %>%
      summarize(pearson_pADT_vs_rADT = cor(real, predicted_sclinear, method = "pearson"),
                pearson_normalizedRNA_vs_rADT = cor(real, rna_normalized, method = "pearson"),
                pearson_rawRNA_vs_rADT = cor(real, rna_raw, method = "pearson"))


d <- d %>% pivot_longer(!gene, values_to = "pearson", names_to = "comparison") %>% arrange(desc(pearson))

d$gene <- factor(d$gene, levels = unique(d$gene))
d$comparison <- factor(d$comparison, levels = c("pearson_pADT_vs_rADT", "pearson_normalizedRNA_vs_rADT", "pearson_rawRNA_vs_rADT"))

p <- ggplot(d, aes(x = gene, y = pearson, fill = comparison)) + geom_bar(stat="identity", position=position_dodge()) +
  theme_bw() + scale_fill_manual(values = kelly()[4:length(kelly())]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
p

ggsave(filename = "./../local/RNA_ADT_correlation_pbmc5k.pdf", plot = p, width = 15, height = 10)
```


#### Pearson correlation by gene

```{r}
PBMC <- read.table("./../local/PBMC5K/bit_table_pbmc5k_2.csv", header = T, sep=',')
PBMC <- na.omit(PBMC)

cors <- PBMC %>% group_by(gene) %>% dplyr::summarise(correlation = cor(real, predicted_sclinear))
ggplot(cors, aes(x=reorder(gene, correlation), y= correlation,fill=correlation)) + geom_bar(stat="identity", col="black") + coord_flip() +
    theme_classic2() + scale_fill_gradientn(colours = inferno(11)) + ylab("Pearson\n(Real vs ScLinear)") + xlab("Protein") + theme(legend.position = "none") +
  ggtitle("PBMC CITE-seq\n(5k cells)") + theme(plot.title = element_text(hjust = 0.5))

```

#### Predicted vs measured ADT scatter plot

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")
PBMC2 <- filter(PBMC, gene %in% c("CD19","CD3","CD14","CD56", "CD8A", "CD8B"))
ggplot(PBMC2, aes(x=predicted_sclinear, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[5:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - Predicted") + ylab("Real")

```

#### Predicted vs measured ADT scatter plot

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")
PBMC2 <- filter(PBMC, gene %in% c("CD19","CD3","CD14","CD56", "CD8A", "CD8B"))
ggplot(PBMC2, aes(x=rna_normalized, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[5:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - RNA_normalized") + ylab("Real")

```