---
title: "Analysis"
author: "Daniel Hanhart"
date: "3/16/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scLinear)
library(tidyverse)
library(ggplot2)
library(Seurat)
library(svglite)
library(Metrics)
library(pals)
library(ggpubr)
library(DropletUtils)
library(ggrepel)
# quickly test if py modules can be loaded
pipe <- create_adt_predictor()
rm(pipe)

set.seed(42)
```

## Mouse lymphnode spleen

### Create Seurat object from raw data

```{r}
# create seurat object from mls data
## Some of the protein names in mls differ from the ones used in the NeurIPS data (CD8a -> CD8, PD-1 -> CD279). Create name mapping."
# how to compare CD21_CD35 vs CD21 and CD35?
 map_names <- data.frame(
 	mouse = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226_10E5", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R_B220", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90.2", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype_G0114F7", "Isotype_HTK888", "Isotype_MOPC.173", "Isotype_MOPC.21", "Isotype_MPC.11", "Isotype_RTK2071", "Isotype_RTK2758", "Isotype_RTK4174",
 	          "Isotype_RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "Ly108", "Ly49D", "Ly49H", "NK.1.1", "Siglec_H", "TCR.Bchain", "TCR.GD_UC7", "TCR.RD_GL3", "TCR.Va2", "TCR.Va8.3_KT50", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA"),
 	
neurips = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90.2", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype_G0114F7", "Isotype_HTK888", "Isotype_MOPC.173", "Isotype_MOPC.21", "Isotype_MPC.11", "Isotype_RTK2071", "Isotype_RTK2758", "Isotype_RTK4174",
            "Isotype_RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "CD352", "Ly49D", "Ly49H", "CD161", "Siglec_H", "TCR.Bchain", "TCR.GD", "TCR.RD", "TCR.Va2", "TCR.Va8.3", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA")
 )


mls.data <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_LymphNode1_BC1_AB1_count_sample_filtered_feature_bc_matrix.h5")








# 
# # https://www.10xgenomics.com/resources/datasets/Mixture-of-cells-from-mouse-lymph-nodes-and-spleen-stained-with-totalseqc-mouse-universal-cocktail
# mls.data <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_Multiplex_count_raw_feature_bc_matrix.h5")
# 
# ## remove HuMsRt. for human, mouse and rat binding antibodies
# rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^HuMsRt\\.", replacement = "",
#                                                            x = rownames(x = mls.data[["Antibody Capture"]]))
# ## remove HuMs. for human and mouse binding antibodies
# rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^HuMs\\.", replacement = "",
#                                                            x = rownames(x = mls.data[["Antibody Capture"]]))
# ## remove MsRt. for human and mouse binding antibodies
# rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^MsRt\\.", replacement = "",
#                                                            x = rownames(x = mls.data[["Antibody Capture"]]))
# ## remove Ms. for mouse binding antibodies
# rownames(x = mls.data[["Antibody Capture"]]) <- gsub(pattern = "^Ms\\.", replacement = "",
#                                                            x = rownames(x = mls.data[["Antibody Capture"]]))
# 
# rownames(x = mls.data[["Antibody Capture"]]) <- map_names$neurips[ match(rownames(x = mls.data[["Antibody Capture"]]), map_names$mouse)]
# 
# # min features set to 0, to model empty droplet removal correctly
# mls <- Seurat::CreateSeuratObject(counts = mls.data[["Gene Expression"]], min.cells = 0, min.features = 0)
# mls[["ADT"]] <- Seurat::CreateAssayObject(mls.data[["Antibody Capture"]][, colnames(x = mls)])
# Seurat::DefaultAssay(mls) <- "RNA"
# 
# mls$mito_percent <- Seurat::PercentageFeatureSet(object = mls, pattern = "^mt-")
# 
# saveRDS(mls, "./../local/mls_raw.rds")
# rm(mls.data, mls, map_names)
# gc()
```

### Preprocess MLS data

```{r}
mls <- readRDS("./../local/mls_raw.rds")


#scLinear::empty_drops(object = mls, lower = 100, FDR = 0.01, samples = NULL, seed = 42)



mls <- scLinear::prepare_data(mls,
                            remove_doublets = TRUE,
                            low_qc_cell_removal = TRUE,
                            anno_level = c(2,3),
                            samples = NULL,
                            integrate_data = FALSE,
                            remove_empty_droplets = TRUE,
                            lower = 100, FDR = 0.01,
                            annotation_selfCluster = TRUE,
                            resolution = 0.2,
                            seed = 42,
                            species = "Mm")
  

DimPlot(mls, group.by = "seurat_clusters", pt.size = 0.05)


## Mark cells for removal based on the ADT assay. Removed only after RNA based qc and preprocessing!
adt_counts <- GetAssayData(mls, slot = "counts", assay = "ADT")

iso_features <- c("IgG1", "IgG2a", "IgG2b")
all_features <- rownames(mls@assays$ADT)
iso_count <- colSums(x = GetAssayData(object = mls, assay = "ADT", slot = "counts")[iso_features, , drop = FALSE])
total_counts_ADT <- colSums(x = GetAssayData(object = mls, assay = "ADT", slot = "counts")[all_features, , drop = FALSE])

qc.stats <- DropletUtils::cleanTagCounts(adt_counts, controls = iso_features) %>% as.data.frame()
high_control_threshold <- attr(qc.stats$high.controls, "thresholds")

DF <- data.frame(total_counts = total_counts_ADT, iso_counts = iso_count, nFeatures = mls$nFeature_ADT)

p<-ggplot(DF, aes(x=total_counts_ADT, y=nFeatures, color = iso_count)) + 
  geom_point(binaxis='y', stackdir='center') + theme_bw() +
  scale_color_gradient(low = "blue", high = "red", trans = "log10") + 
  guides(colour = guide_colourbar(barheight = unit(8, "cm")))
p

DF2 <- DF %>% pivot_longer(!nFeatures, values_to = "counts", names_to = "names")
p <- ggplot(DF2, aes(x = counts, fill = names)) + geom_histogram() +
  theme_bw() +
  scale_fill_manual(values = kelly()[3:length(kelly())]) +
  facet_wrap(~names , ncol = 1, scales = "free") +
  geom_vline(aes(xintercept = xintercept, color = Lines), data.frame(xintercept = high_control_threshold["higher"], Lines = "upper_iso_count_threshold") , linetype = "dashed")
p

## Mark the cells that should be filtered based on adt values (removed after RNA based pre-processing!)
mls@meta.data$adt_filter <- (qc.stats$discard) 



##  remove adt based filtered cells 
# remove cells than have less than 50 
mls@meta.data$adt_filter  <- (mls@meta.data$adt_filter | mls@meta.data$nFeature_ADT < max(mls@meta.data$nFeature_ADT)/2)
print("Number of cells filtered based on ADT data")
table(mls@meta.data$adt_filter )
mls <- subset(mls, subset = adt_filter == FALSE)
## Normalize ADT data
mls <- Seurat::NormalizeData(mls, normalization.method = "CLR", margin = 2, assay = "ADT")
saveRDS(mls, "./../local/mls_prepared2.rds")
#rm(mls,all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT, qc.stats, p, adt_counts, DF, DF2, n_adt_filter)
gc()

# saveRDS(mls, "./../local/mls_prepared.rds")
# rm(mls)
# gc()
```

### Take a look into the ADT data

```{r}
# mls <- readRDS("./../local/mls_prepared.rds")
# 
# adt_counts <- GetAssayData(mls, slot = "counts", assay = "ADT")
# 
# iso_features <- c("Isotype-RTK2071", "Isotype-G0114F7", "Isotype-RTK2758", "Isotype-RTK4174", "Isotype-HTK888")
# all_features <- rownames(mls@assays$ADT)
# iso_count <- colSums(x = GetAssayData(object = mls, assay = "ADT", slot = "counts")[iso_features, , drop = FALSE])
# total_counts_ADT <- colSums(x = GetAssayData(object = mls, assay = "ADT", slot = "counts")[all_features, , drop = FALSE])
# 
# 
# qc.stats <- DropletUtils::cleanTagCounts(adt_counts, controls = iso_features) %>% as.data.frame()
# 
# high_control_threshold <- attr(qc.stats$high.controls, "thresholds")
# 
# DF <- data.frame(total_counts = total_counts_ADT, iso_counts = iso_count, nFeatures = mls$nFeature_ADT)
# 
# p<-ggplot(DF, aes(x=total_counts_ADT, y=nFeatures, color = iso_count)) + 
#   geom_point(binaxis='y', stackdir='center') + theme_bw() +
#   scale_color_gradient(low = "blue", high = "red", trans = "log10") + 
#   guides(colour = guide_colourbar(barheight = unit(8, "cm")))
# p
# 
# 
# ## 
# DF2 <- DF %>% pivot_longer(!nFeatures, values_to = "counts", names_to = "names")
# p <- ggplot(DF2, aes(x = counts, fill = names)) + geom_histogram() +
#   theme_bw() +
#   scale_fill_manual(values = kelly()[3:length(kelly())]) +
#   facet_wrap(~names , ncol = 1, scales = "free") +
#   geom_vline(aes(xintercept = xintercept, color = Lines), data.frame(xintercept = high_control_threshold["higher"], Lines = "upper_iso_count_threshold") , linetype = "dashed")
# 
# p
# 
# ## Normalize ADT data
# mls <- Seurat::NormalizeData(mls, normalization.method = "CLR", margin = 2, assay = "ADT")
# 
# saveRDS(mls, "./../local/mls_prepared_2.rds")
# 
# rm(mls, adt_counts, DF, DF2, p, qc.stats, all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT)
# gc()
```
### UMAP

```{r}
mls <- readRDS("./../local/mls_prepared_2.rds")

p <- DimPlot(mls, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
  scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300", "#69009c")) + theme_bw()
p
DimPlot(mls, group.by = "seurat_clusters", pt.size = 0.05)


## because there are clearly distinct clusters annotated to the same cell type, The specific markers for each cluster are calculated.

# Seurat::Idents(mls) <- mls$seurat_clusters
# 
# cluster_markers <- Seurat::FindAllMarkers(mls)
# 
# a <- cluster_markers %>% dplyr::filter(p_val_adj < 0.05) %>%  group_by(cluster) %>% slice_min(order_by = p_val_adj - abs(avg_log2FC), n = 3, with_ties = FALSE)
# 
# saveRDS(cluster_markers, "./../local/mls_cluster_markers.rds")


ggsave(filename = "./../local/umap_mls.pdf", plot = p, width = 9, height = 6, units = "cm")
rm(p, mls)
gc()
```

### predict mls with scLinear

```{r}
stop()
mls <- readRDS("./../local/mls_prepared_2.rds")

## make a conversion between mouse and human genes
library("biomaRt")
mouse_genes <- rownames(mls@assays$RNA)
#mouse_genes <- mouse_genes[1:4]

human_gene_names <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
mouse_gene_names = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = mouse_genes , mart = mouse_gene_names, attributesL = c("hgnc_symbol"), martL = human_gene_names, uniqueRows=T)

# filter out only genes that are used in the predictor
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")
genesV2 <- genesV2 %>% dplyr::filter(HGNC.symbol %in% pipe$gex_names$to_list())
# We lose quite a bit of genes in the conversion, but the remaining 11959 should be enough for the prediction

# remove genes that are not maped 1:1
keep_1 <- genesV2 %>% group_by(HGNC.symbol) %>% summarise(n=n()) %>% dplyr::filter(n == 1) %>% pull(HGNC.symbol)
keep_2 <- genesV2 %>% group_by(MGI.symbol) %>% summarise(n=n()) %>% dplyr::filter(n == 1) %>% pull(MGI.symbol)

genesV2 <- genesV2 %>% dplyr::filter(HGNC.symbol %in%  keep_1 & MGI.symbol %in% keep_2 )

gexp_mls <- Seurat::GetAssayData( mls@assays$RNA, slot = "counts")
gexp_mls <- gexp_mls[(rownames(gexp_mls) %in% genesV2$MGI.symbol), ]
# We lose quite a bit of genes in the conversion, but the remaining 10658 should be enough for the prediction.
# match gene names

gexp_mls <- gexp_mls %>% as.data.frame() %>% rownames_to_column("MGI.symbol") %>% inner_join(genesV2, by = "MGI.symbol") %>% dplyr::select(!MGI.symbol) %>% column_to_rownames("HGNC.symbol") %>% as.matrix() %>% Matrix::Matrix(sparse = TRUE)
  
gexp_mls_assay <- Seurat::CreateAssayObject(counts = gexp_mls)

## load pretrained predictor model scLinear
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")
assay_predicted <- adt_predict(pipe, gexp_mls_assay, normalize = TRUE)

#eval_res <- evaluate_predictor(pipe, mls@assays$RNA, mls@assays$ADT)

mls[["predicted_adt_sclinear"]] <- assay_predicted

saveRDS(mls, "./../local/mls_predicted.rds")
rm(assay_predicted, eval_res,pipe)
gc()
```

### bit table (RNA / protein predicted vs protein real)

```{r}
mls <- readRDS("./../local/mls_predicted.rds")

adt_real <- as.matrix(mls@assays[["ADT"]]@data)
adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
adt_predicted_sclinear <- as.matrix(mls@assays[["predicted_adt_sclinear"]]@data)
adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_sclinear")


# resolve one (neurips: CD8) to many (mouse: CD8a, CD8b) mapping for CD8

## map adt name to many gexp names. each gene compared to adt name.
adt_names <- "CD8"
genes <- c("CD8a", "CD8b")
tmp <- adt_predicted_sclinear[adt_predicted_sclinear$gene == adt_names,]
adt_predicted_sclinear <- adt_predicted_sclinear[!(adt_predicted_sclinear$gene == adt_names),]
for (g in genes){
  tmp_2 <- tmp
  tmp_2 $gene <- g
  adt_predicted_sclinear <- rbind(adt_predicted_sclinear, tmp_2)
}

meta <- mls@meta.data %>% rownames_to_column("cell") %>% dplyr::select("cell","cell_type_2","cell_type_3")

DF <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell"))  %>% full_join(meta, by = c("cell"))

write.table(DF, file = "./../local/MouseLymphNodes/bit_table_mls_2.csv", sep = ",", col.names = TRUE, row.names = FALSE)
```



```{r, fig.width= 15, fig.height = 10}
DF <- read.table("./../local/MouseLymphNodes/bit_table_mls_2.csv", header = T, sep=',')
d <- DF %>% na.omit() %>% group_by(gene)  %>%
      summarize(pearson_pADT_vs_rADT = cor(real, predicted_sclinear, method = "pearson"))


d <- d %>% pivot_longer(!gene, values_to = "pearson", names_to = "comparison") %>% arrange(desc(pearson))

d$gene <- factor(d$gene, levels = unique(d$gene))
d$comparison <- factor(d$comparison, levels = c("pearson_pADT_vs_rADT"))

p <- ggplot(d, aes(x = gene, y = pearson, fill = comparison)) + geom_bar(stat="identity", position=position_dodge()) +
  theme_bw() + scale_fill_manual(values = kelly()[4:length(kelly())]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
p

ggsave(filename = "./../local/RNA_ADT_correlation_mls.pdf", plot = p, width = 15, height = 10)
```


#### Pearson correlation by gene

```{r, fig.width= 20/3, fig.height = 29/3}
mls_bit_table <- read.table("./../local/MouseLymphNodes/bit_table_mls_2.csv", header = T, sep=',')
mls_bit_table <- na.omit(mls_bit_table)

cors <- mls_bit_table %>% group_by(gene) %>% dplyr::summarise(correlation = cor(real, predicted_sclinear))%>% arrange(desc(correlation)) 
cors$gene <- factor(cors$gene, levels = cors$gene)

ggplot(cors, aes(x=reorder(gene, correlation), y= correlation,fill=correlation)) + geom_bar(stat="identity", col="black") + coord_flip() +
    theme_classic2() + scale_fill_gradientn(colours = inferno(11)) + ylab("Pearson\n(Real vs ScLinear)") + xlab("Protein") + theme(legend.position = "none") +
    ggtitle("MLS CITE-seq\n(30k cells)") + theme(plot.title = element_text(hjust = 0.5)) +
    theme( axis.text.y = element_text(size = 5)) 

#ggsave("./../local/panels/Final/Supplementary/1/A.MLS_Pearson_correlation.r1.pdf" , width = 20/3, height = 29/2, units = 'cm')
#cors <- cors %>% arrange(desc(correlation))


cors$name <- ""
cors$name[cors$gene %in% c("CD3", "CD19", "CD27", "CD11b")] <-  as.vector(cors$gene[cors$gene %in% c("CD3", "CD19", "CD27", "CD11b")])
pos <- position_jitter(width = 0.3, height = 0, seed = 6)
ggplot(cors, aes(x="", y = correlation, fill = "", label = name)) +
  geom_boxplot(show.legend = FALSE) + 
  theme_bw() +
  geom_jitter(position=pos, size = 0.8) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = "#56B4E9") +
  ylab("Pearson\n(Real vs ScLinear)") + xlab("") +
  theme(legend.position = "none") +
  geom_label_repel(position = pos,
                  colour = "black", fill = "white", segment.colour="black",
                  min.segment.length = 0,
                  show.legend = TRUE, size = 2,
                  box.padding = 0.25, point.padding = 0) +
  theme(axis.ticks.x = element_blank(),
         axis.text.y = element_text(size = 8))

ggsave("./../local/panels/Final/Main/2/A.MLS_Pearson_correlation_boxplot.r1.pdf" , width = 20/3, height = 29/3, units = 'cm')

```


#### Predicted vs measured ADT scatter plot

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")
mls_bit_table2 <- filter(mls_bit_table, gene %in% c("CD11b", "CD19","CD27", "CD3"))
mls_bit_table2$gene <- factor(mls_bit_table2$gene, levels = c("CD11b", "CD19","CD3", "CD27"))
ggplot(mls_bit_table2, aes(x=predicted_sclinear, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[4:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - ADT") + ylab("Real") + xlim(0, NA) + ylim(0, NA)

```