---
title: "Analysis"
author: "Daniel Hanhart"
date: "3/16/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scLinear)
library(tidyverse)
library(ggplot2)
library(Seurat)
library(svglite)
library(Metrics)
library(pals)
library(ggpubr)
library(DropletUtils)
library(Matrix)
library(grid)
library(ggrepel)
source("helper_functions.R")
# quickly test if py modules can be loaded
pipe <- create_adt_predictor()
rm(pipe)

set.seed(42)
```

## Mouse lymphnode spleen

### Create Seurat object from raw data

```{r}
# create seurat object from mls data
## Some of the protein names in mls differ from the ones used in the NeurIPS data (CD8a -> CD8, PD-1 -> CD279). Create name mapping."
 map_names <- data.frame(
 	mouse = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226_10E5", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R_B220", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90.2", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype_G0114F7", "Isotype_HTK888", "Isotype_MOPC.173", "Isotype_MOPC.21", "Isotype_MPC.11", "Isotype_RTK2071", "Isotype_RTK2758", "Isotype_RTK4174",
 	          "Isotype_RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "Ly108", "Ly49D", "Ly49H", "NK.1.1", "Siglec_H", "TCR.Bchain", "TCR.GD_UC7", "TCR.RD_GL3", "TCR.Va2", "TCR.Va8.3_KT50", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA"),
 	
neurips = c("CCR3", "CD103", "CD106", "CD107a", "CD115", "CD11a", "CD11b", "CD11c", "CD120b", "CD127", "CD134", "CD137", "CD138", "CD150", "CD155", "CD160", "CD163", "CD169", "CD170", "CD172a", "CD182", "CD183", "CD185", "CD186", "CD19", "CD1d", "CD2", "CD20", "CD200", "CD200R", "CD205", "CD21_CD35", "CD22", "CD223", "CD226", "CD23", "CD24", "CD25", "CD27", "CD272", "CD274", "CD278", "CD279", "CD29", "CD3", "CD301b", "CD304", "CD31", "CD317", "CD357", "CD366", "CD371", "CD38", "CD4", "CD40", "CD41", "CD43", "CD44", "CD45", "CD45.2", "CD45R", "CD45RB", "CD48", "CD49a", "CD49b", "CD49d", "CD49f", "CD5", "CD51", "CD54", "CD55", "CD61", "CD62L", "CD63", "CD69", "CD71", "CD73", "CD81", "CD86", "CD88", "CD8a", "CD8b", "CD9", "CD90", "CD93", "CD94", "CD98", "F4_80", "FceRIa", "FR4", "I.A_I.E", "IgD", "integrin.b7", "Isotype-G0114F7", "Isotype-HTK888", "Isotype-MOPC.173", "Isotype-MOPC.21", "Isotype-MPC.11", "Isotype-RTK2071", "Isotype-RTK2758", "Isotype-RTK4174",
            "Isotype-RTK4530", "KLRG1", "Ly.49A", "Ly.6A_E", "Ly.6C", "Ly.6G", "CD352", "Ly49D", "Ly49H", "CD161", "Siglec_H", "TCR.Bchain", "TCR.GD", "TCR.RD", "TCR.Va2", "TCR.Va8.3", "TCR.Vb5.1_5.2", "TCR.Vb8.1_8.2", "TER.119", "Tim.4", "VISTA")
 )


# source: https://www.10xgenomics.com/resources/datasets/Mixture-of-cells-from-mouse-lymph-nodes-and-spleen-stained-with-totalseqc-mouse-universal-cocktail

mls.data <- list()
mls.data[["LymphNode1"]] <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_LymphNode1_BC1_AB1_count_sample_filtered_feature_bc_matrix.h5")
mls.data[["LymphNode2"]] <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_LymphNode2_BC2_AB2_count_sample_filtered_feature_bc_matrix.h5")
mls.data[["Spleen1"]] <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_Spleen1_BC3_AB3_count_sample_filtered_feature_bc_matrix.h5")
mls.data[["Spleen2"]] <- Seurat::Read10X_h5("./../local/MouseLymphNodes/4plex_mouse_LymphNode_Spleen_TotalSeqC_multiplex_Spleen2_BC4_AB4_count_sample_filtered_feature_bc_matrix.h5")

mls_list <- list()
for(n in names(mls.data)){
  tmp <- mls.data[[n]]
  ## remove HuMsRt. for human, mouse and rat binding antibodies
  rownames(x = tmp[["Antibody Capture"]]) <- gsub(pattern = "^HuMsRt\\.", replacement = "",
                                                             x = rownames(x = tmp[["Antibody Capture"]]))
  ## remove HuMs. for human and mouse binding antibodies
  rownames(x = tmp[["Antibody Capture"]]) <- gsub(pattern = "^HuMs\\.", replacement = "",
                                                             x = rownames(x = tmp[["Antibody Capture"]]))
  ## remove MsRt. for human and mouse binding antibodies
  rownames(x = tmp[["Antibody Capture"]]) <- gsub(pattern = "^MsRt\\.", replacement = "",
                                                             x = rownames(x = tmp[["Antibody Capture"]]))
  ## remove Ms. for mouse binding antibodies
  rownames(x = tmp[["Antibody Capture"]]) <- gsub(pattern = "^Ms\\.", replacement = "",
                                                             x = rownames(x = tmp[["Antibody Capture"]]))
  rownames(x = tmp[["Antibody Capture"]]) <- map_names$neurips[ match(rownames(x = tmp[["Antibody Capture"]]), map_names$mouse)]
  
  # min features set to 0, to model empty droplet removal correctly
  mls <- Seurat::CreateSeuratObject(counts = tmp[["Gene Expression"]], min.cells = 10, min.features = 1)
  mls[["ADT"]] <- Seurat::CreateAssayObject(tmp[["Antibody Capture"]][, colnames(x = mls)])
  Seurat::DefaultAssay(mls) <- "RNA"
  mls$mito_percent <- Seurat::PercentageFeatureSet(object = mls, pattern = "^mt-")
  mls$sample <- n
  mls_list[[n]] <- mls
  rm(tmp, mls)
  gc()
}


saveRDS(mls_list, "./../local/mls1.rds")
rm(mls.data, mls_list, map_names, mls, n)
gc()

```

### Preprocess MLS data

```{r}
mls_list <- readRDS("./../local/mls1.rds")

for(n in names(mls_list)){
  print(paste0("Sample : ", n))
  
  mls <- mls_list[[n]]
  mls <- adt_qc(mls, iso_features = c("Isotype-G0114F7", "Isotype-HTK888", "Isotype-MOPC.173", "Isotype-MOPC.21", "Isotype-MPC.11", "Isotype-RTK2071", "Isotype-RTK2758", "Isotype-RTK4174", "Isotype-RTK4530"))
  mls_list[[n]] <- mls
}

## merge the 4 different samples
mls <- merge(mls_list[[1]], y = mls_list[2:length(names(mls_list))])
table(mls@meta.data[["sample"]])

mls <- scLinear::prepare_data(mls,
                            remove_doublets = TRUE,
                            low_qc_cell_removal = TRUE,
                            anno_level = c(2,3),
                            samples = "sample",
                            integrate_data = TRUE,
                            remove_empty_droplets = FALSE,
                            annotation_selfCluster = TRUE,
                            resolution = 0.8,
                            seed = 42,
                            species = "Mm")
  

##  remove adt based filtered cells 
# remove cells than have less than 50% percent of nFeatures_adt
print("Number of cells filtered based on ADT data")
table(mls@meta.data$adt_filter )
mls <- subset(mls, subset = adt_filter == FALSE)

## Normalize ADT data
mls <- Seurat::NormalizeData(mls, normalization.method = "CLR", margin = 2, assay = "ADT")

DimPlot(mls, group.by = "seurat_clusters", pt.size = 0.05)
DimPlot(mls, group.by = "sample", pt.size = 0.05)
DimPlot(mls, group.by = "cell_type_2", pt.size = 0.05)


DimPlot(mls, group.by = "seurat_clusters", cells = colnames(mls)[mls@meta.data$cell_type_2 == "Unassigned"]
)
Seurat::Idents(mls) <- mls@meta.data$cell_type_2
Seurat::DotPlot(mls, assay = "ADT", features = c("CD3", "CD4", "CD19", "CD11b"))

## unassigned label clearly belongs to the T-cell cluster

mls@meta.data$cell_type_2[(mls@meta.data$cell_type_2 == "Unassigned")] <- "T"
DimPlot(mls, group.by = "cell_type_2", pt.size = 0.05)

saveRDS(mls, "./../local/mls_prepared2.rds")
rm(mls,all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT, qc.stats, p, adt_counts, DF, DF2)
gc()


```


### UMAP

```{r}
mls <- readRDS("./../local/mls_prepared2.rds")

p <- DimPlot(mls, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
  scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300", "#69009c")) + theme_bw()
p
DimPlot(mls, group.by = "seurat_clusters", pt.size = 0.05)

ggsave(filename = "./../local/umap_mls.pdf", plot = p, width = 9, height = 6, units = "cm")
rm(p, mls)
gc()
```


#### Train on mouse samples

```{r}
get_bit_table <- function(object_train, object_test){
  pipe <- create_adt_predictor()
  pipe <- fit_predictor(pipe = pipe,
              gexp_train = object_train@assays[["RNA"]],
              adt_train = object_train@assays[["ADT"]],
              normalize_gex = TRUE,
              normalize_adt = TRUE)
  
  predicted_assay <- adt_predict(pipe = pipe, gexp = object_test@assays[["RNA"]], normalize = TRUE, slot = "counts")
  object_test[["predicted_adt"]] <- predicted_assay

  adt_real <- as.matrix(object_test@assays[["ADT"]]@data)
  adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
  adt_predicted_sclinear <- as.matrix(object_test@assays[["predicted_adt"]]@data)
  adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_adt")
  
  
  meta <- object_test@meta.data %>% rownames_to_column("cell")
  
  bit_table <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>% full_join(meta, by = c("cell"))
  
  return(bit_table)
}


mls <- readRDS("./../local/mls_prepared2.rds")
mls_list <- Seurat::SplitObject(mls, split.by = "sample")


# train vs test LymphNode
train_set <- mls_list$LymphNode1
test_set <- mls_list$LymphNode2
LymphNode1_vs_LymphNode2 <- get_bit_table(train_set, test_set) %>% mutate(train_test = "LymphNode2")
train_set <- mls_list$LymphNode2
test_set <- mls_list$LymphNode1
LymphNode2_vs_LymphNode1 <- get_bit_table(train_set, test_set) %>% mutate(train_test = "LymphNode1")
# train vs test Spleen
train_set <- mls_list$Spleen1
test_set <- mls_list$Spleen2
Spleen1_vs_Spleen2 <- get_bit_table(train_set, test_set) %>% mutate(train_test = "Spleen2")
train_set <- mls_list$Spleen2
test_set <- mls_list$Spleen1
Spleen2_vs_Spleen1 <- get_bit_table(train_set, test_set) %>% mutate(train_test = "Spleen1")



DF <- do.call(rbind, list(LymphNode1_vs_LymphNode2,LymphNode2_vs_LymphNode1,Spleen1_vs_Spleen2, Spleen2_vs_Spleen1))

saveRDS(DF, "./../local/mls_other_species.rds")

DF <- readRDS("./../local/mls_other_species.rds")
DF <- DF[!grepl(pattern = "Isotype", DF$gene),]

#b <- a %>% dplyr::select(cell, gene, real, predicted_adt, cell_type_2) %>% group_by(cell)  %>% summarise(Pearson = cor(real, predicted_adt)) %>% ungroup() %>% summarise(Pearson = mean(Pearson))
df <- DF %>% dplyr::select(cell, gene, real, predicted_adt, cell_type_2, train_test) %>% group_by(gene, train_test)  %>% summarise(Pearson = cor(real, predicted_adt)) %>% ungroup()
#pivot_longer(cols = c("real","predicted_adt"))





my_pal <- c("#A2A187", "#525240", "#68C6A4", "#03878F")
df$name <- ""
#df$name[cors$gene %in% c("CD3", "CD19", "CD27", "CD11b")] <-  as.vector(df$gene[df$gene %in% c("CD3", "CD19", "CD27", "CD11b")])
pos <- position_jitter(width = 0.3, height = 0, seed = 6)
ggplot(df, aes(x="", y = Pearson, fill = train_test, label = name)) +
  geom_boxplot(show.legend = FALSE) + 
  theme_bw() +
  geom_jitter(position=pos, size = 0.8) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = my_pal) +
  ylab("Pearson\n(Real vs ScLinear)") + xlab("") +
  theme(legend.position = "none") +
  geom_label_repel(position = pos,
                  colour = "black", fill = "white", segment.colour="black",
                  min.segment.length = 0,
                  show.legend = TRUE, size = 1.5,
                  box.padding = 0.25, point.padding = 0) +
  theme(axis.ticks.x = element_blank(),
         axis.text.y = element_text(size = 8),
        panel.spacing = unit(0.05, "lines"),
        strip.text.x = element_text(size = 3)) +
  facet_grid(~train_test)




#ggsave("./../local/panels/Final/Main/2/A.MLS_Pearson_correlation_boxplot_mls_trained.r1.pdf" , width = 20/2, height = 29/3, units = 'cm')
```

