---
title: "Analysis"
author: "Daniel Hanhart"
date: "3/16/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scLinear)
library(tidyverse)
library(ggplot2)
library(Seurat)
library(svglite)
library(Metrics)
library(pals)
library(ggpubr)
library(DropletUtils)

set.seed(42)
```

## PBMC5K

### Create Seurat object from raw data

```{r}
# create seurat object from pbmc5k data
## Some of the protein names in pbmc5k differ from the ones used in the NeurIPS data (CD8a -> CD8, PD-1 -> CD279). Create name mapping."
map_names <- data.frame(
	pbmc5k = c("CD3", "CD4", "CD8a", "CD11b", "CD14", "CD15", "CD16", "CD19", "CD20", "CD25", "CD27", "CD28", "CD34", "CD45RA", "CD45RO", "CD56", "CD62L", "CD69", "CD80", "CD86", "CD127", "CD137", "CD197", "CD274", "CD278", "CD335", "PD-1", "HLA-DR", "TIGIT", "IgG1", "IgG2a", "IgG2b"),
	neurips = c("CD3", "CD4", "CD8", "CD11b", "CD14", "CD15", "CD16", "CD19", "CD20", "CD25", "CD27", "CD28", "CD34", "CD45RA", "CD45RO", "CD56", "CD62L", "CD69", "CD80", "CD86", "CD127", "CD137", "CD197", "CD274", "CD278", "CD335", "CD279", "HLA-DR", "TIGIT", "IgG1", "IgG2a", "IgG2b")
)

pbmc5k.data <- Seurat::Read10X_h5("./../local/PBMC5K/5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5")
rownames(x = pbmc5k.data[["Antibody Capture"]]) <- gsub(pattern = "_[control_]*TotalSeqB", replacement = "",
                                                           x = rownames(x = pbmc5k.data[["Antibody Capture"]]))
rownames(x = pbmc5k.data[["Antibody Capture"]]) <- map_names$neurips[ match(rownames(x = pbmc5k.data[["Antibody Capture"]]), map_names$pbmc5k)]

pbmc5k <- Seurat::CreateSeuratObject(counts = pbmc5k.data[["Gene Expression"]], min.cells = 0, min.features = 0)
pbmc5k[["ADT"]] <- Seurat::CreateAssayObject(pbmc5k.data[["Antibody Capture"]][, colnames(x = pbmc5k)])
Seurat::DefaultAssay(pbmc5k) <- "RNA"

pbmc5k$mito_percent <- Seurat::PercentageFeatureSet(object = pbmc5k, pattern = "^MT-")
pbmc5k$ribo_percent <- Seurat::PercentageFeatureSet(object = pbmc5k, pattern = "^RP[SL]")

saveRDS(pbmc5k, "./../local/pbmc5k_filtered.rds")
rm(pbmc5k.data, pbmc5k, map_names)
gc()
```

### Preprocess PBMC5K data

```{r}
pbmc5k <- readRDS("./../local/pbmc5k_filtered.rds")

## Perform basc preprocessing steps including: lognormalization, removal of doublets, removal of low quality cells, clustering, cell-type annotation
pbmc5k <- scLinear::prepare_data(pbmc5k,
                                   remove_doublets = TRUE,
                                   low_qc_cell_removal = TRUE,
                                   anno_level = c(2,3),
                                   samples = NULL,
                                   integrate_data = FALSE,
                                   remove_empty_droplets = FALSE,
                                   lower = 100, FDR = 0.01,
                                   annotation_selfCluster = TRUE,
                                   resolution = 0.8,
                                   seed = 42)

saveRDS(pbmc5k, "./../local/PBMC5K_prepared.rds")
rm(pbmc5k)
gc()
```
### Take a look into the ADT data
```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared.rds")

adt_counts <- GetAssayData(pbmc5k, slot = "counts", assay = "ADT")

iso_features <- c("IgG1", "IgG2a", "IgG2b")
all_features <- rownames(pbmc5k@assays$ADT)
iso_count <- colSums(x = GetAssayData(object = pbmc5k, assay = "ADT", slot = "counts")[iso_features, , drop = FALSE])
total_counts_ADT <- colSums(x = GetAssayData(object = pbmc5k, assay = "ADT", slot = "counts")[all_features, , drop = FALSE])


qc.stats <- DropletUtils::cleanTagCounts(adt_counts, controls = iso_features) %>% as.data.frame()

high_control_threshold <- attr(qc.stats$high.controls, "thresholds")

DF <- data.frame(total_counts = total_counts_ADT, iso_counts = iso_count, nFeatures = pbmc5k$nFeature_ADT)

p<-ggplot(DF, aes(x=total_counts_ADT, y=nFeatures, color = iso_count)) + 
  geom_point(binaxis='y', stackdir='center') + theme_bw() +
  scale_color_gradient(low = "blue", high = "red", trans = "log10") + 
  guides(colour = guide_colourbar(barheight = unit(8, "cm")))
p


## 
DF2 <- DF %>% pivot_longer(!nFeatures, values_to = "counts", names_to = "names")
p <- ggplot(DF2, aes(x = counts, fill = names)) + geom_histogram() +
  theme_bw() +
  scale_fill_manual(values = kelly()[3:length(kelly())]) +
  facet_wrap(~names , ncol = 1, scales = "free") +
  geom_vline(aes(xintercept = xintercept, color = Lines), data.frame(xintercept = high_control_threshold["higher"], Lines = "upper_iso_count_threshold") , linetype = "dashed")

p

## Normalize ADT data
pbmc5k <- Seurat::NormalizeData(pbmc5k, normalization.method = "CLR", margin = 2, assay = "ADT")

saveRDS(pbmc5k, "./../local/PBMC5K_prepared_2.rds")

rm(pbmc5k, adt_counts, DF, DF2, p, qc.stats, all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT)
gc()
```
### UMAP

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared_2.rds")

p <- DimPlot(pbmc5k, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
  scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300")) + theme_bw()
p
ggsave(filename = "./../local/umap_pbmc5k.pdf", plot = p, width = 9, height = 6, units = "cm")
rm(p, pbmc5k)
gc()
```

### predict PBMC5K with scLinear

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_prepared_2.rds")

## load pretrained predictor model scLinear
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")
assay_predicted <- adt_predict(pipe, pbmc5k@assays$RNA, normalize = TRUE)

eval_res <- evaluate_predictor(pipe, pbmc5k@assays$RNA, pbmc5k@assays$ADT)

pbmc5k[["predicted_adt_sclinear"]] <- assay_predicted

saveRDS(pbmc5k, "./../local/PBMC5K_predicted.rds")
rm(assay_predicted, eval_res,pipe)
gc()
```

### bit table (RNA / protein predicted vs protein real)

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_predicted.rds")

adt_real <- as.matrix(pbmc5k@assays[["ADT"]]@data)
adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
adt_predicted_sclinear <- as.matrix(pbmc5k@assays[["predicted_adt_sclinear"]]@data)
adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_sclinear")

#### add raw RNA
rna_raw <- as.matrix(pbmc5k@assays[["RNA"]]@counts)
rna_raw <- rna_raw %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_raw")
#### post processing of RNA data to make it comparable to ADT data
rna_raw$gene2 <- rna_raw$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_raw$gene2[rna_raw$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_raw$gene2[rna_raw$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_raw$gene2[rna_raw$gene %in% c("ITGAM")] <- "CD11b"
rna_raw$gene2[rna_raw$gene %in% c("FUT4")] <- "CD15"
rna_raw$gene2[rna_raw$gene %in% c("FCGR3A")] <- "CD16"
rna_raw$gene2[rna_raw$gene %in% c("MS4A1")] <- "CD20"
rna_raw$gene2[rna_raw$gene %in% c("IL2RA")] <- "CD25"
rna_raw$gene2[rna_raw$gene %in% c("NCAM1")] <- "CD56"
rna_raw$gene2[rna_raw$gene %in% c("SELL")] <- "CD62L"
rna_raw$gene2[rna_raw$gene %in% c("IL7R")] <- "CD127"
rna_raw$gene2[rna_raw$gene %in% c("TNFR5F9")] <- "CD137"
rna_raw$gene2[rna_raw$gene %in% c("CCR7")] <- "CD197"
rna_raw$gene2[rna_raw$gene %in% c("ICOS")] <- "CD278"
rna_raw$gene2[rna_raw$gene %in% c("NCR1")] <- "CD335"
rna_raw$gene2[rna_raw$gene %in% c("PDCD1")] <- "CD279"
## three genes part of the HLA-DR complex are measures (take mean at the end)
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRA")] <- "HLA-DR"
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRB1")] <- "HLA-DR"
rna_raw$gene2[rna_raw$gene %in% c("HLA-DRB5")] <- "HLA-DR"


### PTPRC -> CD45RA / CD45RO (ignored for nown not clear to which isoform the gene counts should be compared to!)
#-
rna_raw <- rna_raw %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_raw") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_raw <- rna_raw %>% group_by(cell, gene) %>% summarise(rna_raw = mean(rna_raw))

#### add normalized RNA (normalized the same way as used for prediction)
rna_real <- as.matrix(scLinear::gexp_normalize(pbmc5k@assays[["RNA"]]@counts))
rna_real <- rna_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_normalized")
#### post processing of RNA data to make it comparable to ADT data
rna_real$gene2 <- rna_real$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_real$gene2[rna_real$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_real$gene2[rna_real$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_real$gene2[rna_real$gene %in% c("ITGAM")] <- "CD11b"
rna_real$gene2[rna_real$gene %in% c("FUT4")] <- "CD15"
rna_real$gene2[rna_real$gene %in% c("FCGR3A")] <- "CD16"
rna_real$gene2[rna_real$gene %in% c("MS4A1")] <- "CD20"
rna_real$gene2[rna_real$gene %in% c("IL2RA")] <- "CD25"
rna_real$gene2[rna_real$gene %in% c("NCAM1")] <- "CD56"
rna_real$gene2[rna_real$gene %in% c("SELL")] <- "CD62L"
rna_real$gene2[rna_real$gene %in% c("IL7R")] <- "CD127"
rna_real$gene2[rna_real$gene %in% c("TNFR5F9")] <- "CD137"
rna_real$gene2[rna_real$gene %in% c("CCR7")] <- "CD197"
rna_real$gene2[rna_real$gene %in% c("ICOS")] <- "CD278"
rna_real$gene2[rna_real$gene %in% c("NCR1")] <- "CD335"
rna_real$gene2[rna_real$gene %in% c("PDCD1")] <- "CD279"
## three genes part of the HLA-DR complex are measures (take mean at the end)
rna_real$gene2[rna_real$gene %in% c("HLA-DRA")] <- "HLA-DR"
rna_real$gene2[rna_real$gene %in% c("HLA-DRB1")] <- "HLA-DR"
rna_real$gene2[rna_real$gene %in% c("HLA-DRB5")] <- "HLA-DR"

### PTPRC -> CD45RA / CD45RO (ignored for nown not clear to which isoform the gene counts should be compared to!)
#-
rna_real <- rna_real %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_normalized") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_real <- rna_real %>% group_by(cell, gene) %>% summarise(rna_normalized = mean(rna_normalized))

meta <- pbmc5k@meta.data %>% rownames_to_column("cell") %>% dplyr::select("cell","cell_type_2","cell_type_3")

DF <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>%  full_join(rna_real, by = c("gene", "cell")) %>% full_join(rna_raw, by = c("gene", "cell")) %>% full_join(meta, by = c("cell"))

write.table(DF, file = "./../local/PBMC5K/bit_table_pbmc5k_2.csv", sep = ",", col.names = TRUE, row.names = FALSE)
```


### bit table (RNA / protein predicted vs protein real) (take 2)

```{r}
pbmc5k <- readRDS("./../local/PBMC5K_predicted.rds")

adt_real <- as.matrix(pbmc5k@assays[["ADT"]]@data)
adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
adt_predicted_sclinear <- as.matrix(pbmc5k@assays[["predicted_adt_sclinear"]]@data)
adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_sclinear")

#### raw RNA counts
rna_raw <- as.matrix(pbmc5k@assays[["RNA"]]@counts)
rna_raw <- rna_raw %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_raw")
#### post processing of RNA data to make it comparable to ADT data

#### add normalized RNA (normalized the same way as used for prediction)
rna_normalized <- as.matrix(scLinear::gexp_normalize(pbmc5k@assays[["RNA"]]@counts))
rna_normalized <- rna_normalized %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_normalized")
#### post processing of RNA data to make it comparable to ADT data

map <- list(
    list(gexp = c("ITGAM"), adt = c("CD11b")),
    list(gexp = c("FUT4"), adt = c("CD15")),
    list(gexp = c("FCGR3A"), adt = c("CD16")),
    list(gexp = c("MS4A1"), adt = c("CD20")),
    list(gexp = c("IL2RA"), adt = c("CD25")),
    list(gexp = c("NCAM1"), adt = c("CD56")),
    list(gexp = c("SELL"), adt = c("CD62L")),
    list(gexp = c("IL7R"), adt = c("CD127")),
    list(gexp = c("TNFR5F9"), adt = c("CD137")),
    list(gexp = c("CCR7"), adt = c("CD197")),
    list(gexp = c("ICOS"), adt = c("CD278")),
    list(gexp = c("NCR1"), adt = c("CD335")),
    list(gexp = c("PDCD1"), adt = c("CD279")),
    list(gexp = c("HLA-DRA", "HLA-DRB1", "HLA-DRB5"), adt = c("HLA-DR")),
    list(gexp = c("CD3E", "CD3D", "CD3G"), adt = c("CD3")),
    list(gexp = c("CD8A", "CD8B"), adt = c("CD8")),
    list(gexp = c("PTPRC"), adt = c("CD45RA", "CD45RO"))
  )


for (m in map){
  print(paste0("gexp: ", paste0(m$gexp, collapse = ","), "       adt: ", paste0(m$adt, collapse = ",")))
  
  gexp_names <- m$gexp
  adt_names <- m$adt
  
  if((length(gexp_names) == 1) & (length(adt_names) == 1)){
    ## change rna name to adt names
    rna_raw$gene[rna_raw$gene == c(gexp_names)] <- adt_names
    rna_normalized$gene[rna_normalized$gene == c(gexp_names)] <- adt_names
  }else{
    if((length(gexp_names) > 1) & (length(adt_names) == 1)){
      ## map adt name to many gexp names. each gene compared to adt name.
      genes <- gexp_names
      tmp <- adt_real[adt_real$gene == adt_names,]
      adt_real <- adt_real[!(adt_real$gene == adt_names),]
      for (g in genes){
        tmp_2 <- tmp
        tmp_2 $gene <- g
        adt_real <- rbind(adt_real, tmp_2)
      }
      genes <- gexp_names
      tmp <- adt_predicted_sclinear[adt_predicted_sclinear$gene == adt_names,]
      adt_predicted_sclinear <- adt_predicted_sclinear[!(adt_predicted_sclinear$gene == adt_names),]
      for (g in genes){
        tmp_2 <- tmp
        tmp_2 $gene <- g
        adt_predicted_sclinear <- rbind(adt_predicted_sclinear, tmp_2)
      }
      
    }else{
        if((length(gexp_names) == 1) & (length(adt_names) > 1)){
          genes <- adt_names
          tmp <- rna_raw[rna_raw$gene == gexp_names,]
          rna_raw <- rna_raw[!(rna_raw$gene == gexp_names),]
          for (g in genes){
            tmp_2 <- tmp
            tmp_2 $gene <- g
            rna_raw <- rbind(rna_raw, tmp_2)
          }

          genes <- adt_names
          tmp <- rna_normalized[rna_normalized$gene == gexp_names,]
          rna_normalized <- rna_normalized[!(rna_normalized$gene == gexp_names),]
          for (g in genes){
            tmp_2 <- tmp
            tmp_2 $gene <- g
            rna_normalized <- rbind(rna_normalized, tmp_2)
          }
          
        }
    }
  }
  
}


## remove not usefull genes
genes_to_keep <- unique(c(unlist(map, recursive = TRUE), adt_real$gene, adt_predicted_sclinear$gene))
rna_raw <- rna_raw %>% dplyr:: filter(gene %in% genes_to_keep)
rna_normalized <- rna_normalized %>% dplyr:: filter(gene %in% genes_to_keep)

meta <- pbmc5k@meta.data %>% rownames_to_column("cell") %>% dplyr::select("cell","cell_type_2","cell_type_3")

DF <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>%  full_join(rna_normalized, by = c("gene", "cell")) %>% full_join(rna_raw, by = c("gene", "cell")) %>% full_join(meta, by = c("cell"))

DF <- DF %>% arrange(gene)

DF$gene <- factor(DF$gene, levels = unique(DF$gene))

write.table(DF, file = "./../local/PBMC5K/bit_table_pbmc5k_2.csv", sep = ",", col.names = TRUE, row.names = FALSE)
```


```{r, fig.width= 15, fig.height = 10}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")

DF <- read.table("./../local/PBMC5K/bit_table_pbmc5k_2.csv", header = T, sep=',')
d <- DF %>% na.omit() %>% group_by(gene)  %>%
      summarize(pearson_pADT_vs_rADT = cor(real, predicted_sclinear, method = "pearson"),
                pearson_normalizedRNA_vs_rADT = cor(real, rna_normalized, method = "pearson"),
                pearson_rawRNA_vs_rADT = cor(real, rna_raw, method = "pearson"))


d <- d %>% pivot_longer(!gene, values_to = "pearson", names_to = "comparison") %>% arrange(desc(pearson))

d$gene <- factor(d$gene, levels = unique(d$gene))
d$comparison <- factor(d$comparison, levels = c("pearson_pADT_vs_rADT", "pearson_normalizedRNA_vs_rADT", "pearson_rawRNA_vs_rADT"))

p <- ggplot(d, aes(x = gene, y = pearson, fill = comparison)) + geom_bar(stat="identity", position=position_dodge()) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_manual(values = rev(c(my_pal))[5:length(my_pal)], 
                    labels=c('Predicted ADT vs meassured ADT', 'Normalized RNA expression vs meassured ADT', 'Raw RNA expression vs meassured ATD')) +
  labs(fill = "Pearson Correlation") 
  
p

Â¨#ggsave(filename = "./../local/RNA_ADT_correlation_pbmc5k.pdf", plot = p, width = 15, height = 10)
```


#### Pearson correlation by gene

```{r}
PBMC <- read.table("./../local/PBMC5K/bit_table_pbmc5k_2.csv", header = T, sep=',')
PBMC <- na.omit(PBMC)

cors <- PBMC %>% group_by(gene) %>% dplyr::summarise(correlation = cor(real, predicted_sclinear))
ggplot(cors, aes(x=reorder(gene, correlation), y= correlation,fill=correlation)) + geom_bar(stat="identity", col="black") + coord_flip() +
    theme_classic2() + scale_fill_gradientn(colours = inferno(11)) + ylab("Pearson\n(Real vs ScLinear)") + xlab("Protein") + theme(legend.position = "none") +
  ggtitle("PBMC CITE-seq\n(5k cells)") + theme(plot.title = element_text(hjust = 0.5))

```

#### Predicted vs measured ADT scatter plot

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")
PBMC2 <- filter(PBMC, gene %in% c("CD19","CD3","CD14","CD56", "CD8A", "CD8B"))
ggplot(PBMC2, aes(x=predicted_sclinear, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[5:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - Predicted") + ylab("Real")

```

#### Predicted vs measured ADT scatter plot

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")
PBMC2 <- filter(PBMC, gene %in% c("CD19","CD3","CD14","CD56", "CD8A", "CD8B"))
ggplot(PBMC2, aes(x=rna_normalized, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[5:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - RNA_normalized") + ylab("Real")

```