---
title: "Analysis"
author: "Daniel Hanhart"
date: "3/16/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scLinear)
library(tidyverse)
library(ggplot2)
library(Seurat)
library(svglite)
library(Metrics)
library(pals)
library(ggpubr)
library(Matrix)
library(grid)
library(ggrepel)
# quickly test if py modules can be loaded
pipe <- create_adt_predictor()
rm(pipe)

marker_genes_rna <- readRDS("./../local/marker_genes_rna.rds")
marker_genes_adt_predicted <- readRDS("./../local/marker_genes_adt_predicted.rds")

set.seed(42)
```

## PBMC10K

### Create Seurat object from raw data

```{r}
## data from: https://satijalab.org/seurat/archive/v3.2/multimodal_vignette.html
# create seurat object from pbmc10k data

## Some of the protein names differ from the ones used in the NeurIPS data (CD8a -> CD8, PD-1 -> CD279). Create name mapping."
map_names <- data.frame(
	pbmc10k = c("CD3", "CD4", "CD8a", "CD14", "CD15", "CD16", "CD56", "CD19", "CD25", "CD45RA", "CD45RO", "PD-1", "TIGIT", "CD127", "IgG2a", "IgG1", "IgG2b"),
	neurips = c("CD3", "CD4", "CD8", "CD14", "CD15", "CD16", "CD56", "CD19", "CD25", "CD45RA", "CD45RO", "CD279", "TIGIT", "CD127", "IgG2a", "IgG1", "IgG2b")
)

pbmc10k.data <- Seurat::Read10X_h5("./../local/PBMC/pbmc_10k_protein_v3_filtered_feature_bc_matrix.h5")
rownames(x = pbmc10k.data[["Antibody Capture"]]) <- gsub(pattern = "_[control_]*TotalSeqB", replacement = "",
                                                           x = rownames(x = pbmc10k.data[["Antibody Capture"]]))
rownames(x = pbmc10k.data[["Antibody Capture"]]) <- map_names$neurips[ match(rownames(x = pbmc10k.data[["Antibody Capture"]]), map_names$pbmc10k)]


pbmc10k <- Seurat::CreateSeuratObject(counts = pbmc10k.data[["Gene Expression"]], min.cells = 1, min.features = 1)
pbmc10k[["ADT"]] <- Seurat::CreateAssayObject(pbmc10k.data[["Antibody Capture"]][, colnames(x = pbmc10k)])
Seurat::DefaultAssay(pbmc10k) <- "RNA"

pbmc10k$mito_percent <- Seurat::PercentageFeatureSet(object = pbmc10k, pattern = "^MT-")

saveRDS(pbmc10k, "./../local/pbmc10k_filtered.rds")
rm(pbmc10k.data, map_names, pbmc10k)
gc()
```


### Preprocess PBMC10K data

```{r}
pbmc10k <- readRDS("./../local/pbmc10k_filtered.rds")

## Mark cells for removal based on the ADT assay. Removed only after RNA based qc and preprocessing!
#pbmc10k <- readRDS("./../local/PBMC10K_prepared.rds")
adt_counts <- GetAssayData(pbmc10k, slot = "counts", assay = "ADT")

iso_features <- c("IgG1", "IgG2a", "IgG2b")
all_features <- rownames(pbmc10k@assays$ADT)
iso_count <- colSums(x = GetAssayData(object = pbmc10k, assay = "ADT", slot = "counts")[iso_features, , drop = FALSE])
total_counts_ADT <- colSums(x = GetAssayData(object = pbmc10k, assay = "ADT", slot = "counts")[all_features, , drop = FALSE])

qc.stats <- DropletUtils::cleanTagCounts(adt_counts, controls = iso_features) %>% as.data.frame()
high_control_threshold <- attr(qc.stats$high.controls, "thresholds")


DF <- data.frame(total_counts = total_counts_ADT, iso_counts = iso_count, nFeatures = pbmc10k$nFeature_ADT)

p<-ggplot(DF, aes(x=total_counts_ADT, y=nFeatures, color = iso_count)) + 
  geom_point(binaxis='y', stackdir='center') + theme_bw() +
  scale_color_gradient(low = "blue", high = "red", trans = "log10") + 
  guides(colour = guide_colourbar(barheight = unit(8, "cm")))
p

DF2 <- DF %>% pivot_longer(!nFeatures, values_to = "counts", names_to = "names")
p <- ggplot(DF2, aes(x = counts, fill = names)) + geom_histogram() +
  theme_bw() +
  scale_fill_manual(values = kelly()[3:length(kelly())]) +
  facet_wrap(~names , ncol = 1, scales = "free") +
  geom_vline(aes(xintercept = xintercept, color = Lines), data.frame(xintercept = high_control_threshold["higher"], Lines = "upper_iso_count_threshold") , linetype = "dashed")
p

## Mark the cells that should be filtered based on adt values (removed after RNA based pre-processing!)
pbmc10k@meta.data$adt_filter <- (qc.stats$discard) 



## Perform basc preprocessing steps including: lognormalization, removal of doublets, removal of low quality cells, clustering, cell-type annotation
pbmc10k <- scLinear::prepare_data(pbmc10k,
                                   remove_doublets = TRUE,
                                   low_qc_cell_removal = TRUE,
                                   anno_level = c(1,2,3,4),
                                   samples = NULL,
                                   integrate_data = FALSE,
                                   remove_empty_droplets = FALSE,
                                   lower = 100, FDR = 0.01,
                                   annotation_selfCluster = TRUE,
                                   resolution = 0.8,
                                   seed = 42)

DimPlot(pbmc10k, group.by = "seurat_clusters")
DimPlot(pbmc10k, group.by = "cell_type_2")


##  remove adt based filtered cells 
# remove cells than have less than 50% percent of nFeatures_adt
pbmc10k@meta.data$adt_filter  <- (pbmc10k@meta.data$adt_filter | pbmc10k@meta.data$nFeature_ADT < max(pbmc10k@meta.data$nFeature_ADT)/2)
print("Number of cells filtered based on ADT data")
table(pbmc10k@meta.data$adt_filter )
pbmc10k <- subset(pbmc10k, subset = adt_filter == FALSE)

## Normalize ADT data
pbmc10k <- Seurat::NormalizeData(pbmc10k, normalization.method = "CLR", margin = 2, assay = "ADT")
saveRDS(pbmc10k, "./../local/PBMC10K_prepared2.rds")
rm(pbmc10k,all_features, high_control_threshold, iso_count, iso_features, total_counts_ADT, qc.stats, p, adt_counts, DF, DF2, n_adt_filter)
gc()
```

### UMAP

```{r}
pbmc10k <- readRDS("./../local/PBMC10K_prepared2.rds")

p <- DimPlot(pbmc10k, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
  scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300")) + theme_bw()
p
ggsave(filename = "./../local/umap_pbmc10k.pdf", plot = p, width = 9, height = 6, units = "cm")
rm(p, pbmc10k)
gc()
```

### predict PBMC10K with scLinear

```{r}
pbmc10k <- readRDS("./../local/PBMC10K_prepared2.rds")

## load pretrained predictor model scLinear
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")
assay_predicted <- adt_predict(pipe, pbmc10k@assays$RNA, normalize = TRUE)

eval_res <- evaluate_predictor(pipe, pbmc10k@assays$RNA, pbmc10k@assays$ADT)

pbmc10k[["predicted_adt_sclinear"]] <- assay_predicted

saveRDS(pbmc10k, "./../local/PBMC10K_predicted.rds")
rm(assay_predicted, eval_res,pipe, pbmc10k)
gc()
```

### bit table (RNA / protein predicted vs protein real)

```{r}
pbmc10k <- readRDS("./../local/PBMC10K_predicted.rds")

adt_real <- as.matrix(pbmc10k@assays[["ADT"]]@data)
adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
adt_predicted_sclinear <- as.matrix(pbmc10k@assays[["predicted_adt_sclinear"]]@data)
adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_sclinear")

#### add raw RNA
rna_raw <- as.matrix(pbmc10k@assays[["RNA"]]@counts)
rna_raw <- rna_raw %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_raw")
#### post processing of RNA data to make it comparable to ADT data
rna_raw$gene2 <- rna_raw$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_raw$gene2[rna_raw$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_raw$gene2[rna_raw$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_raw$gene2[rna_raw$gene %in% c("FUT4")] <- "CD15"
rna_raw$gene2[rna_raw$gene %in% c("FCGR3A")] <- "CD16"
rna_raw$gene2[rna_raw$gene %in% c("NCAM1")] <- "CD56"
rna_raw$gene2[rna_raw$gene %in% c("IL2RA")] <- "CD25"
rna_raw$gene2[rna_raw$gene %in% c("PDCD1")] <- "CD279"
rna_raw$gene2[rna_raw$gene %in% c("IL7R")] <- "CD127"
### PTPRC -> CD45RA / CD45RO (ignored for now not clear to which isoform the gene counts should be compared to!)
#-
rna_raw <- rna_raw %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_raw") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_raw <- rna_raw %>% group_by(cell, gene) %>% summarise(rna_raw = mean(rna_raw))

#### add normalized RNA (normalized the same way as used for prediction)
rna_real <- as.matrix(scLinear::gexp_normalize(pbmc10k@assays[["RNA"]]@counts))
rna_real <- rna_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "rna_normalized")
#### post processing of RNA data to make it comparable to ADT data
rna_real$gene2 <- rna_real$gene
## assign all components of complex to same complex eg CD3 D/E/G -> CD3
rna_real$gene2[rna_real$gene %in% c("CD3E", "CD3D", "CD3G")] <- "CD3"
rna_real$gene2[rna_real$gene %in% c("CD8A", "CD8B")] <- "CD8"
## change names (different naming convention between RNA and ADT assay)
rna_real$gene2[rna_real$gene %in% c("FUT4")] <- "CD15"
rna_real$gene2[rna_real$gene %in% c("FCGR3A")] <- "CD16"
rna_real$gene2[rna_real$gene %in% c("NCAM1")] <- "CD56"
rna_real$gene2[rna_real$gene %in% c("IL2RA")] <- "CD25"
rna_real$gene2[rna_real$gene %in% c("PDCD1")] <- "CD279"
rna_real$gene2[rna_real$gene %in% c("IL7R")] <- "CD127"
### PTPRC -> CD45RA / CD45RO (ignored for nown not clear to which isoform the gene counts should be compared to!)
#-
rna_real <- rna_real %>% mutate(gene = gene2) %>% dplyr::select("gene", "cell", "rna_normalized") %>% dplyr:: filter(gene %in% unique(adt_real$gene))
## combine complex expression by taking the mean expression value
rna_real <- rna_real %>% group_by(cell, gene) %>% summarise(rna_normalized = mean(rna_normalized))

meta <- pbmc10k@meta.data %>% rownames_to_column("cell") %>% dplyr::select("cell","cell_type_1","cell_type_2","cell_type_3","cell_type_4")

DF <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>%  full_join(rna_real, by = c("gene", "cell")) %>% full_join(rna_raw, by = c("gene", "cell")) %>% full_join(meta, by = c("cell"))

write.table(DF, file = "./../local/PBMC/bit_table_pbmc10k_2.csv", sep = ",", col.names = TRUE, row.names = FALSE)

```

### RNA, ADT, ADT predicted correlation comparison

```{r}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")

DF <- read.table("./../local/PBMC/bit_table_pbmc10k_2.csv", header = T, sep=',')
df <- DF %>% na.omit() %>% group_by(gene)  %>%
      summarize(pearson_pADT_vs_rADT = cor(real, predicted_sclinear, method = "pearson"),
                pearson_normalizedRNA_vs_rADT = cor(real, rna_normalized, method = "pearson"),
                pearson_rawRNA_vs_rADT = cor(real, rna_raw, method = "pearson"))


df <- df %>% pivot_longer(!gene, values_to = "pearson", names_to = "comparison") %>% arrange(desc(pearson))

df$gene <- factor(df$gene, levels = unique(df$gene))
df$comparison <- factor(df$comparison, levels = c("pearson_pADT_vs_rADT", "pearson_normalizedRNA_vs_rADT", "pearson_rawRNA_vs_rADT"))

p <- ggplot(df, aes(x = gene, y = pearson, fill = comparison)) + geom_bar(stat="identity", position=position_dodge()) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_manual(values = rev(c(my_pal))[5:length(my_pal)], 
                    labels=c('Predicted ADT vs meassured ADT', 'Normalized RNA expression vs meassured ADT', 'Raw RNA expression vs meassured ATD')) +
  labs(fill = "Comparison") +
  ylab("Pearson correlation") +
  xlab("Gene") +
  ggtitle("")
p

#ggsave("./../local/panels/Final/Supplementary/1/A.PBMC10K_Protein_RNA_correlation.r1.pdf", width = 20, height = 29/4, units = 'cm')
#ggsave(filename = "./../local/RNA_ADT_correlation_pbmc5k.pdf", plot = p, width = 15, height = 10)
```

##### Heatmap

```{r, fig.width= 20/2, fig.height = 29/2}
my_pal <- c("#E7E7E0","#A2A187", "#525240", "#03878F", "#F2B531","#ED5958","#68C6A4", "#113245")

DF <- read.table("./../local/PBMC/bit_table_pbmc10k_2.csv", header = T, sep=',')
d <- DF %>% na.omit() %>% group_by(gene)  %>%
      summarize(pearson_pADT_vs_rADT = cor(real, predicted_sclinear, method = "pearson"),
                pearson_normalizedRNA_vs_rADT = cor(real, rna_normalized, method = "pearson"),
                pearson_rawRNA_vs_rADT = cor(real, rna_raw, method = "pearson")) %>% column_to_rownames("gene")

d <- d %>% arrange(desc(pearson_pADT_vs_rADT),desc(pearson_normalizedRNA_vs_rADT), desc(pearson_rawRNA_vs_rADT))

p <- ComplexHeatmap::pheatmap(d, scale = "row", 
                   cluster_rows = FALSE, cluster_cols = FALSE,
                   display_numbers = round(as.matrix(d),2),
                   labels_col = c('Predicted ADT\nvs\nMeasured ADT\n\n', 'Normalized RNA Expression\nvs\nMeasured ADT\n\n', 'Raw RNA Expression\nvs\nMeasured ATD\n\n'),
                   legend = TRUE,
                   angle_col = c("0"),
                   color = rev(brewer.rdbu(11))[2:10],
                   number_color = "black", border_color = "white", border = TRUE, border_gp = gpar(col = "black", lty = 1, lwd = 1.5),
                   number_format ="%.2f"  , fontsize_number = 6,
                   column_names_side = c("top"), row_names_side = c("left"),
                   fontsize = 6,
                   fontsize_row = 6,
                   fontsize_col = 5,
                   column_names_max_height = unit(100, "cm"),
                   heatmap_legend_param = list(title = "z score"))
#hcl.colors(50, "sunsetdark", rev=F)
ComplexHeatmap::draw(p, padding = unit(c(0, 0, 20, 0), "mm"))

pdf("./../local/panels/Final/Supplementary/1/C.PBMC10K_Protein_RNA_correlation_heatmap.r1.pdf", width = (20/2)/2.54, height = (29/3 * 2)/2.54)
  ComplexHeatmap::draw(p, padding = unit(c(2, 2, 20, 2), "mm"))
dev.off()
```




##### Boxplots
```{r, fig.width= 20/2, fig.height = 29/2}
DF <- read.table("./../local/PBMC/bit_table_pbmc10k_2.csv", header = T, sep=',')

df <- DF %>% na.omit()  %>% group_by(gene)  %>%
      summarize('scLinear vs ADT' = cor(real, predicted_sclinear, method = "pearson"),
                'RNA vs ADT' = cor(real, rna_normalized, method = "pearson"))

df <- df %>%pivot_longer(!gene, values_to = "Pearson", names_to = "Comparison")
df$Comparison <- factor(df$Comparison, levels = rev(sort(unique(df$Comparison))))

df$gene_comp <- paste0(df$gene, "_", df$Comparison)


my_pal <- c("#68C6A4", "#A2A187")
df <- df %>% arrange(gene, Comparison)
df$name <- df$gene
df$name[!(df$gene %in% c("CD3", "CD19", "CD14", "CD56"))] <- ""
pos <- position_jitter(width = 0.2, height = 0, seed = 3)
ggplot(df, aes(x=Comparison, y = Pearson, fill = Comparison, label = name)) +
  geom_boxplot(show.legend = FALSE) + 
  theme_bw() +
  geom_point(position=pos, size = 0.9) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = my_pal) +
  ylab("Pearson") + xlab("") +
  theme(legend.position = "none") +
  geom_label_repel(position = pos,
                  colour = "black", fill = "white", segment.colour="black",
                  min.segment.length = 1,
                  show.legend = TRUE, size = 1.5,
                  box.padding = 0.25, point.padding = 0) +
  theme(axis.ticks.x = element_blank(),
         axis.text.y = element_text(size = 8),
        panel.spacing = unit(0.05, "lines"),
        strip.text.x = element_text(size = 7)) +
  geom_line(aes(group = gene), position = pos, size = 0.25 , color = "darkblue") 

ggsave("./../local/panels/Final/Main/1/B.PBMC10K_Pearson_RNA_correlation.v2.pdf", width = 20/2, height = 29/2, units = 'cm')
```



```{r, fig.width=20/3, fig.height=29/4}
#########
### E ###
#########
PBMC <- read.table("./../local/PBMC/bit_table_pbmc10k_2.csv", header = T, sep=',')
PBMC <- PBMC %>% dplyr::select(-c(rna_normalized, rna_raw))
PBMC <- na.omit(PBMC)

cors <- PBMC %>% group_by(gene) %>% dplyr::summarise(correlation = cor(real, predicted_sclinear))
ggplot(cors, aes(x=reorder(gene, correlation), y= correlation,fill=correlation)) + geom_bar(stat="identity", col="black") + coord_flip() +
    theme_classic2() + scale_fill_gradientn(colours = inferno(11)) + ylab("Pearson\n(Real vs ScLinear)") + xlab("Protein") + theme(legend.position = "none") +
  ggtitle("PBMC CITE-seq\n(10k cells)") + theme(plot.title = element_text(hjust = 0.5))

ggsave("./../local/panels/Final/Main/1/E.PBMC10K_Pearson.v2.pdf", width = 20/3, height = 29/3, units = 'cm')
```


```{r, fig.width=20/3*2, fig.height=29/3}
#ggsave("panels/Final/Main/1/E.Pearson.v2.pdf", width = 20/3, height = 29/3, units = 'cm')

#########
### F ###
#########
PBMC2 <- filter(PBMC, gene %in% c("CD19","CD3","CD14","CD56"))
ggplot(PBMC2, aes(x=predicted_sclinear, y=real, col=cell_type_2)) + geom_point(alpha = 0.8, size=0.5) + facet_wrap(~gene, scales = "free") +
  scale_color_manual(values = my_pal[5:8]) + theme_classic2() + theme(legend.title = element_blank()) +
  geom_smooth(method=lm, se=FALSE, col=my_pal[2], alpha = 0.5) + geom_rug() +
  xlab("ScLinear - Predicted") + ylab("Real")

ggsave("./../local/panels/Final/Main/1/F.PBMC10K_Markers.pdf", width = 20/3*2, height = 29/3, units = "cm")

```



### Expression plots
```{r}
# pbmc10k <- readRDS("./../local/PBMC10K_predicted.rds")
# 
# p <- DimPlot(pbmc10k, group.by = "cell_type_2", pt.size = 0.05) + ggtitle("Cell types") +
#   scale_color_manual(values = c("#BE0032","#008856","#0067A5", "#F3C300")) + theme_bw()
# p
# 
# create_feature_plot <- function(obj, feature){
#   DefaultAssay(pbmc10k) <- "ADT"
#   p1 <- Seurat::FeaturePlot(obj, features = feature, slot = "data") +   
#               theme(legend.title = element_blank()) + scale_fill_gradientn(colours = inferno(11)) +
#               theme(legend.position = "none")
#   
#   DefaultAssay(pbmc10k) <- "predicted_adt_sclinear"
#   p2 <- Seurat::FeaturePlot(obj, features = feature, slot = "data") +   
#               theme(legend.title = element_blank()) + scale_fill_gradientn(colours = inferno(11)) +
#               theme(legend.position = "none")
#   
#   p3 <- ggpubr::ggarrange(p1, p2, nrow = 1)
#   return(p3)
# }
# 
# create_feature_plot(obj = pbmc10k, feature = "CD14")
# create_feature_plot(obj = pbmc10k, feature = "CD19")
# create_feature_plot(obj = pbmc10k, feature = "CD3")
# create_feature_plot(obj = pbmc10k, feature = "CD56")
# create_feature_plot(obj = pbmc10k, feature = "CD279")
# create_feature_plot(obj = pbmc10k, feature = "CD25")
# create_feature_plot(obj = pbmc10k, feature = "TIGIT")
# 
# create_feature_plot(obj = pbmc10k, feature = "CD279")
# 
# create_feature_plot(obj = pbmc10k, feature = "CD279")
# 
# create_feature_plot(obj = pbmc10k, feature = "CD279")
# 
# 
# a <- pbmc10k
# a@assays[["predicted_adt_sclinear"]] <- NULL
# a@assays[["ADT"]] <- NULL
# #a@assays[["RNA"]] <- NULL
# a <- a %>% RunPCA(a)
# 
# 
# create_feature_plot(obj = a, feature = "CD279")

```



### feature impact analysis PBMC10K

```{r}
if(FALSE){
  pbmc10k <- readRDS("./../local/PBMC10K_predicted.rds")
  
  pipe <- create_adt_predictor()
  pipe <- load_pretrained_model(pipe, model = "all")
  
  feature_importance <- function(pipe, gexp, slot = "counts", normalize = TRUE){
      gexp_matrix <- Seurat::GetAssayData(gexp, slot = slot)
    
      if(normalize){
      ## normalize data GEX
      sce <- SingleCellExperiment::SingleCellExperiment(list(counts = gexp_matrix))
      clusters <- scran::quickCluster(sce)
      sce <- scran::computeSumFactors(sce, clusters=clusters)
      sce <- scuttle::logNormCounts(sce, pseudo.count = 1, center.size.factors = FALSE, log = FALSE)
      gexp_matrix <- sce@assays@data@listData[["normcounts"]]
      gexp_matrix <- base::log1p(gexp_matrix)
    }
    gexp_matrix <- Matrix::t(gexp_matrix)
    gexp_matrix_py <- reticulate::r_to_py(as.matrix(gexp_matrix))
    feature_imp <- pipe$feature_importance(gexp_matrix_py, gex_names = colnames(gexp_matrix))
    colnames(feature_imp) <- pipe$gex_names$to_list()
    rownames(feature_imp) <- pipe$adt_names$to_list()
    
    return(feature_imp)
  }
  
  fi <- feature_importance(pipe, pbmc10k@assays$RNA, slot = "counts", normalize = TRUE)
  
  write.table(fi, "./../local/PBMC/feature_importance_pbmc10k.txt", sep = "\t", row.names = TRUE, col.names = TRUE, quote = FALSE)
  saveRDS(fi, "./../local/PBMC/feature_importance_pbmc10k.rds")
}
```




```{r, fig.width=20/3*2, fig.height=29/3}

#ggsave("panels/Final/Main/2/D.CD19.CD3assays.pdf", height = 29/3, width = 20/3*2, units = "cm")

pbmc <- read.table("./../local/PBMC/feature_importance_pbmc10k.txt", header = T)

pbmc_bit_table <- read.table("./../local/PBMC/bit_table_pbmc10k_2.csv", header = T, sep=',')
pbmc_bit_table <- pbmc_bit_table %>% dplyr::select(-c(rna_normalized, rna_raw))
pbmc_bit_table <- na.omit(pbmc_bit_table)
cors <- pbmc_bit_table %>% group_by(gene) %>% dplyr::summarise(correlation = cor(real, predicted_sclinear))
important_proteins <- cors %>% group_by("gene") %>% slice_max(correlation, n = 2) %>% pull(gene)
important_proteins <- c("CD3", "CD19")

important_genes <- pbmc %>% as.data.frame() %>% rownames_to_column("protein") %>%
  dplyr::filter(protein %in% important_proteins) %>%
  pivot_longer(!protein, values_to = "importance", names_to = "gene") %>%
  group_by(protein) %>% slice_max(importance, n = 20) %>% pull("gene") %>% unique()

important_features_matrix <- pbmc %>% t() %>% as.data.frame() %>% rownames_to_column("gene") %>%
  dplyr::filter(gene %in% important_genes) %>% column_to_rownames("gene") %>% t()

important_features_matrix <- important_features_matrix[rownames(important_features_matrix) %in% important_proteins,]

ComplexHeatmap::pheatmap(important_features_matrix)
```



### Drop out simulation


```{r, fig.width=20/3*2, fig.height=29/3}
set.seed(42)
obj <-  readRDS("./../local/PBMC10K_prepared2.rds")
df <- GetAssayData(obj, assay = "RNA", slot = "counts")
index <- which(df != 0)
drops <- matrix(nrow = nrow(df), ncol = ncol(df), data = 1)
drops[index] <- 0
R <- nrow(df)
C <- ncol(df)
drop_rate <- length(which(drops == 1))/(R*C)
print(paste0("Initial dropout rate: ", drop_rate))

dropout_data_list <- list()

while(drop_rate < 0.99){
    index <- which(drops == 0)
    
    new_drops <- rbinom(length(index), 1, 0.1)
    
    drops[index[which(new_drops == 1)]] <- 1
    drop_rate <- length(which(drops == 1))/(R*C)
    print(paste0("New dropout: ", drop_rate))
    df_drop <- df
    df_drop[which(drops == 1)] <- 0
    dropout_data_list[[as.character(drop_rate)]] <- df_drop
    
}    

dropout_levels <- names(dropout_data_list)
for(i in 1:length(dropout_data_list)){
  check_df <- dropout_data_list[[dropout_levels[i]]]
  print(length(which(check_df == 0))/(R*C))
}

saveRDS(dropout_data_list, "./../local/pbmc10k_dropout_list.rds")
```




### predict PBMC10K with scLinear for drop out

```{r}
get_dropout_rate <- function(object){
  df <- GetAssayData(object, assay = "RNA",  slot = "counts")
  index <- which(df != 0)
  drops <- matrix(nrow = nrow(df), ncol = ncol(df), data = 1)
  drops[index] <- 0
  R <- nrow(df)
  C <- ncol(df)
  drop_rate <- length(which(drops == 1))/(R*C)
  return(drop_rate)
}

get_dropout_res <- function(pipe,object_test , input_type){
  do_rate <- get_dropout_rate(object_test)
  predicted_assay <- adt_predict(pipe = pipe, gexp = object_test@assays[["RNA"]], normalize = TRUE, slot = "counts")
  object_test[["predicted_adt"]] <- predicted_assay

  adt_real <- as.matrix(object_test@assays[["ADT"]]@data)
  adt_real <- adt_real %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "real")
  adt_predicted_sclinear <- as.matrix(object_test@assays[["predicted_adt"]]@data)
  adt_predicted_sclinear <- adt_predicted_sclinear %>% as.data.frame() %>% rownames_to_column("gene") %>% pivot_longer(!gene,names_to = "cell", values_to = "predicted_adt")
  
  
  meta <- object_test@meta.data %>% rownames_to_column("cell")
  
  bit_table <- adt_real %>% full_join(adt_predicted_sclinear, by = c("gene", "cell")) %>% full_join(meta, by = c("cell")) %>% mutate(type = input_type, dropout = do_rate)
  
  
  df <- bit_table %>% dplyr::select(cell, gene, real, predicted_adt, cell_type_2, type, dropout) %>% na.omit() %>% group_by(cell)
  
  return(df)
}


pbmc10k <- readRDS("./../local/PBMC10K_prepared2.rds")
## load pretrained predictor model scLinear
pipe <- create_adt_predictor()
pipe <- load_pretrained_model(pipe, model = "all")


dropout_res_list <- list()
dropout_res_list[["original"]] <- get_dropout_res(pipe, pbmc10k, "original")


dropout_data_list <- readRDS("./../local/pbmc10k_dropout_list.rds")

for(i in names(dropout_data_list)){
  print(i)
  obj <- pbmc10k
  new_assay <- Seurat::CreateAssayObject(counts = dropout_data_list[[i]])
  obj[["RNA"]] <- new_assay
  dropout_res_list[[i]] <- get_dropout_res(pipe, obj, "added_dropout")
  rm(obj)
  gc()
}

dropout_res <- do.call(rbind, dropout_res_list)

saveRDS(dropout_res, "./../local/dropout_res_pbmc10k.rds")

```


```{r,  fig.width=20/2, fig.height=29/2}
pbmc10k <- readRDS("./../local/PBMC10K_predicted.rds")

my_pal <- c("#525240", "#03878F")
dropout_res <- readRDS("./../local/dropout_res_pbmc10k.rds")


DF <- dropout_res%>% group_by(cell, dropout, type) %>% summarise(Pearson = cor(real, predicted_adt))
DF$Pearson <- as.numeric(DF$Pearson )

DF$type [!(DF$type == "original")]<- "added_dropout"

pos <- position_jitter(width = 0, height = 0, seed = 6)
p <- ggplot(DF, aes(x = dropout, y = Pearson, color = type, group = dropout)) +
  geom_boxplot(show.legend = FALSE,  outlier.size = 0.2) + theme_bw() +
  #geom_jitter(position=pos, size = 0.8) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = my_pal) +
  ylab("Pearson\n(Real vs ScLinear)") + xlab("") +
  theme(legend.position = "none") +
   scale_color_manual(values = c(my_pal)) + scale_fill_manual(values = c(my_pal))  +
  theme(legend.position = "none") +
  theme(legend.title = element_blank()) +
  theme(legend.text = element_text(size = 8)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ylab("Pearson") +
  xlab("Dropout rate") + ggtitle("PBMC10K")

p

ggsave(plot = p, "./../local/panels/Final/Main/2/B.PBMC10K_Dropout_boxplot.r1.pdf", width = 20/2, height = 29/3, units = 'cm')
```

```{r}
pbmc10k <- readRDS("./../local/PBMC10K_predicted.rds")

DF <- pbmc10k@meta.data

df <- DF %>% dplyr::select(cell_type_2)
table(df)


plot_nebulosa <- function(object, assay = "RNA", features,slot = "data", joint = FALSE, title_name = "", tool = "nebulosa"){
  Seurat::DefaultAssay(object) <- assay
  if(tool == "nebulosa"){
    p <- Nebulosa::plot_density(object, features = features, slot = slot, joint = joint) + labs(title = title_name, caption = paste0("Assay: ", assay, "; features: ", paste0(features, collapse = ",")))
  }else{
     p <- Seurat::FeaturePlot(object, features = features, slot = slot) + labs(title = title_name, caption = paste0("Assay: ", assay, "; features: ", paste0(features, collapse = ",")))
  }
  return(p)
}

Seurat::DimPlot(pbmc10k, group.by = "seurat_clusters")
Seurat::DimPlot(pbmc10k, group.by = "cell_type_2")


```

### cluster based on adt assays
```{r, fig.widh = 24, fig.height = 12}
# pbmc10k@meta.data$RNA_Clusters <- pbmc10k@meta.data$seurat_clusters
# 
# pbmc10k <- Seurat::FindVariableFeatures(pbmc10k, assay = "ADT")
# pbmc10k <- Seurat::FindVariableFeatures(pbmc10k, assay = "predicted_adt_sclinear")
# 
# pbmc10k <- Seurat::ScaleData(pbmc10k, assay = "ADT")
# pbmc10k <- Seurat::ScaleData(pbmc10k, assay = "predicted_adt_sclinear")
# 
# pbmc10k <- Seurat::RunPCA(pbmc10k, assay = "ADT", reduction.name = "pca_adt")
# pbmc10k <- Seurat::RunPCA(pbmc10k, assay = "predicted_adt_sclinear", reduction.name = "pca_predicted_adt")
# 
# pbmc10k <- Seurat::FindNeighbors(pbmc10k, assay = "ADT", reduction = "pca_adt")
# pbmc10k <- Seurat::FindNeighbors(pbmc10k, assay = "predicted_adt_sclinear", reduction = "pca_predicted_adt")
# 
# pbmc10k <- Seurat::FindClusters(pbmc10k, graph.name = "ADT_snn")
# pbmc10k@meta.data$ADT_Clusters <- pbmc10k@meta.data$seurat_clusters
# 
# pbmc10k <- Seurat::FindClusters(pbmc10k, graph.name = "predicted_adt_sclinear_snn")
# pbmc10k@meta.data$predicted_adt_Clusters <- pbmc10k@meta.data$seurat_clusters
# 
# 
# # a <- Seurat::FindNeighbors(a)
# # a <- Seurat::FindClusters(a)
# pbmc10k <- Seurat::RunUMAP(pbmc10k, assay = "ADT", dims = 1:min(nrow(pbmc10k@assays$ADT), 30), reduction.name = "umap_adt")
# pbmc10k <- Seurat::RunUMAP(pbmc10k, assay = "predicted_adt_sclinear", dims = 1:min(nrow(pbmc10k@assays$predicted_adt_sclinear), 30), reduction.name = "umap_predicted_adt")
# 
# 
# 
# p1 <- Seurat::DimPlot(pbmc10k, group.by = "RNA_Clusters", reduction = "umap")
# p2 <- Seurat::DimPlot(pbmc10k, group.by = "ADT_Clusters", reduction = "umap_adt")
# p3 <- Seurat::DimPlot(pbmc10k, group.by = "predicted_adt_Clusters", reduction = "umap_predicted_adt")
# 
# p11 <- Seurat::DimPlot(pbmc10k, group.by = "cell_type_2", reduction = "umap")
# p22 <- Seurat::DimPlot(pbmc10k, group.by = "cell_type_2", reduction = "umap_adt")
# p33 <- Seurat::DimPlot(pbmc10k, group.by = "cell_type_2", reduction = "umap_predicted_adt")
# 
# (p1 | p2 | p3) / (p11 | p22 | p33)
# 
# adt_markers <- Seurat::FindAllMarkers(pbmc10k, assay = "ADT")
# adt_predicted_markers <- Seurat::FindAllMarkers(pbmc10k, assay = "predicted_adt_sclinear")
# 
# 
# 


```
```{r, fig.widh = 12, fig.height = 6}
# cluster_subtype <- function(object, assay_name = "RNA" ){
#   
#   object <- Seurat::FindVariableFeatures(object, assay = assay_name)
#   object <- Seurat::ScaleData(object, assay = assay_name)
#   object <- Seurat::RunPCA(object, assay = assay_name)
#   object <- Seurat::FindNeighbors(object, assay = assay_name)
#   object <- Seurat::FindClusters(object, graph.name = paste0(assay_name,"_snn"))
#   object <- Seurat::RunUMAP(object, assay = assay_name, dims = 1:min((nrow(object@assays[[assay_name]])-1), 30))
#   return(object)
# }
# 
# 
# sobj_T <- subset(pbmc10k, subset = cell_type_2 == "T")
# sobj_T <- Seurat::DietSeurat(sobj_T)
# a <- cluster_subtype(sobj_T, assay = "ADT")
# 
# DimPlot(a)
# 
# # plot_nebulosa(a, assay = "ADT", features = "CD3", slot = "data")
# # plot_nebulosa(a, assay = "ADT", features = "CD127", slot = "data")
# # plot_nebulosa(a, assay = "ADT", features = "CD8", slot = "data")
# # plot_nebulosa(a, assay = "predicted_adt_sclinear", features = "CD8", slot = "data")
# # plot_nebulosa(a, assay = "ADT", features = "CD25", slot = "data", tool = "seurat")
# # plot_nebulosa(a, assay = "predicted_adt_sclinear", features = "CD25", slot = "data", tool = "seurat")
# plot_nebulosa(a, assay = "ADT", features = "CD45RO", slot = "data", tool = "seurat")
# plot_nebulosa(a, assay = "predicted_adt_sclinear", features = "CD45RO", slot = "data", tool = "seurat")
# plot_nebulosa(a, assay = "ADT", features = "CD45RA", slot = "data", tool = "seurat")
# plot_nebulosa(a, assay = "predicted_adt_sclinear", features = "CD45RA", slot = "data", tool = "seurat")
# 
# Seurat::DotPlot(a, assay = c("ADT"), features = shared_proteins, scale = FALSE)
# Seurat::DotPlot(a, assay = c("predicted_adt_sclinear"), features = shared_proteins, scale = FALSE)

```

### one to one feature comparison

```{r, fig.widh = 12, fig.height = 6}
shared_proteins <- rownames(pbmc10k@assays$ADT)[rownames(pbmc10k@assays$ADT) %in% rownames(pbmc10k@assays$predicted_adt_sclinear)]
  
for(g in shared_proteins){
  p1 <- plot_nebulosa(pbmc10k, assay = "ADT", features = g, slot = "data")
  p2 <- plot_nebulosa(pbmc10k, assay = "predicted_adt_sclinear", features = g, slot = "data")
  print((p1 | p2)) 
}
```
```{r, fig.widh = 12, fig.height = 6}
shared_proteins <- rownames(pbmc10k@assays$ADT)[rownames(pbmc10k@assays$ADT) %in% rownames(pbmc10k@assays$predicted_adt_sclinear)]
  
for(g in shared_proteins){
  p1 <- plot_nebulosa(pbmc10k, assay = "ADT", features = g, slot = "data", tool = "seurat")
  p2 <- plot_nebulosa(pbmc10k, assay = "predicted_adt_sclinear", features = g, slot = "data",  tool = "seurat")
  print((p1 | p2)) 
}
```


### Try multi color scale in umap



```{r}
object <- readRDS("./../local/PBMC10K_predicted.rds")


umap <- object@reductions[["umap"]]

```


```{r}

sessionInfo()

```
